<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Wrtiting ORFs In Files · GeneFinder.jl</title><meta name="title" content="Wrtiting ORFs In Files · GeneFinder.jl"/><meta property="og:title" content="Wrtiting ORFs In Files · GeneFinder.jl"/><meta property="twitter:title" content="Wrtiting ORFs In Files · GeneFinder.jl"/><meta name="description" content="Documentation for GeneFinder.jl."/><meta property="og:description" content="Documentation for GeneFinder.jl."/><meta property="twitter:description" content="Documentation for GeneFinder.jl."/><meta property="og:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/iodocs/"/><meta property="twitter:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/iodocs/"/><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/iodocs/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.svg" alt="GeneFinder.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../naivefinder/">Finding ORFs</a></li><li><a class="tocitem" href="../simplecodingrule/">A Simple Coding Rule</a></li><li class="is-active"><a class="tocitem" href>Wrtiting ORFs In Files</a><ul class="internal"><li><a class="tocitem" href="#Writting-ORFs-into-bioinformatic-formats"><span>Writting ORFs into bioinformatic formats</span></a></li></ul></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Wrtiting ORFs In Files</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Wrtiting ORFs In Files</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h2 id="Writting-ORFs-into-bioinformatic-formats"><a class="docs-heading-anchor" href="#Writting-ORFs-into-bioinformatic-formats">Writting ORFs into bioinformatic formats</a><a id="Writting-ORFs-into-bioinformatic-formats-1"></a><a class="docs-heading-anchor-permalink" href="#Writting-ORFs-into-bioinformatic-formats" title="Permalink"></a></h2><p>This package facilitates the creation of <code>FASTA</code>, <code>BED</code>, and <code>GFF</code> files, specifically extracting Open Reading Frame (ORF) information from <code>BioSequence</code> instances, particularly those of type <code>NucleicSeqOrView{A} where A</code>, and then writing the information into the desired format.</p><p>Functionality:</p><p>The package provides four distinct functions for writing files in different formats:</p><table><tr><th style="text-align: right">Function</th><th style="text-align: right">Description</th></tr><tr><td style="text-align: right"><code>write_orfs_fna</code></td><td style="text-align: right">Writes nucleotide sequences in FASTA format.</td></tr><tr><td style="text-align: right"><code>write_orfs_faa</code></td><td style="text-align: right">Writes amino acid sequences in FASTA format.</td></tr><tr><td style="text-align: right"><code>write_orfs_bed</code></td><td style="text-align: right">Outputs information in BED format.</td></tr><tr><td style="text-align: right"><code>write_orfs_gff</code></td><td style="text-align: right">Generates files in GFF format.</td></tr></table><p>All these functions support processing both <code>BioSequence</code> instances and external <code>FASTA</code> files. In the case of a <code>BioSequence</code> instace into external files, simply provide the path to the <code>FASTA</code> file using a <code>String</code> to the path. To demonstrate the use of the <code>write_*</code> methods with a <code>BioSequence</code>, consider the following example:</p><pre><code class="language-julia hljs">using BioSequences, GeneFinder

# &gt; 180195.SAMN03785337.LFLS01000089 -&gt; finds only 1 gene in Prodigal (from Pyrodigal tests)
seq = dna&quot;AACCAGGGCAATATCAGTACCGCGGGCAATGCAACCCTGACTGCCGGCGGTAACCTGAACAGCACTGGCAATCTGACTGTGGGCGGTGTTACCAACGGCACTGCTACTACTGGCAACATCGCACTGACCGGTAACAATGCGCTGAGCGGTCCGGTCAATCTGAATGCGTCGAATGGCACGGTGACCTTGAACACGACCGGCAATACCACGCTCGGTAACGTGACGGCACAAGGCAATGTGACGACCAATGTGTCCAACGGCAGTCTGACGGTTACCGGCAATACGACAGGTGCCAACACCAACCTCAGTGCCAGCGGCAACCTGACCGTGGGTAACCAGGGCAATATCAGTACCGCAGGCAATGCAACCCTGACGGCCGGCGACAACCTGACGAGCACTGGCAATCTGACTGTGGGCGGCGTCACCAACGGCACGGCCACCACCGGCAACATCGCGCTGACCGGTAACAATGCACTGGCTGGTCCTGTCAATCTGAACGCGCCGAACGGCACCGTGACCCTGAACACAACCGGCAATACCACGCTGGGTAATGTCACCGCACAAGGCAATGTGACGACTAATGTGTCCAACGGCAGCCTGACAGTCGCTGGCAATACCACAGGTGCCAACACCAACCTGAGTGCCAGCGGCAATCTGACCGTGGGCAACCAGGGCAATATCAGTACCGCGGGCAATGCAACCCTGACTGCCGGCGGTAACCTGAGC&quot;</code></pre><p>Once a <code>BioSequence</code> object has been instantiated, the <code>write_orfs_fna</code> function proves useful for generating a <code>FASTA</code> file containing the nucleotide sequences of the ORFs. Notably, the <code>write_orfs*</code> methods support either an <code>IOStream</code> or an <code>IOBuffer</code> as an output argument, allowing flexibility in directing the output either to a file or a buffer. In the following example, we demonstrate writing the output directly to a file.</p><pre><code class="language-julia hljs">outfile = &quot;LFLS01000089.fna&quot;

open(outfile, &quot;w&quot;) do io
    write_orfs_fna(seq, io, NaiveFinder())
end</code></pre><pre><code class="language-bash hljs">cat LFLS01000089.fna

&gt;seq id=01 start=29 stop=40 strand=+ frame=2 score=0.0
ATGCAACCCTGA
&gt;seq id=02 start=137 stop=145 strand=+ frame=2 score=0.0
ATGCGCTGA
&gt;seq id=03 start=164 stop=184 strand=+ frame=2 score=0.0
ATGCGTCGAATGGCACGGTGA
&gt;seq id=04 start=173 stop=184 strand=+ frame=2 score=0.0
ATGGCACGGTGA
&gt;seq id=05 start=236 stop=241 strand=+ frame=2 score=0.0
ATGTGA
&gt;seq id=06 start=248 stop=268 strand=+ frame=2 score=0.0
ATGTGTCCAACGGCAGTCTGA
&gt;seq id=07 start=362 stop=373 strand=+ frame=2 score=0.0
ATGCAACCCTGA
&gt;seq id=08 start=470 stop=496 strand=+ frame=2 score=0.0
ATGCACTGGCTGGTCCTGTCAATCTGA
&gt;seq id=09 start=551 stop=574 strand=+ frame=2 score=0.0
ATGTCACCGCACAAGGCAATGTGA
&gt;seq id=10 start=569 stop=574 strand=+ frame=2 score=0.0
ATGTGA
&gt;seq id=11 start=581 stop=601 strand=+ frame=2 score=0.0
ATGTGTCCAACGGCAGCCTGA
&gt;seq id=12 start=695 stop=706 strand=+ frame=2 score=0.0
ATGCAACCCTGA</code></pre><p>This could also be done to writting a <code>FASTA</code> file with the nucleotide sequences of the ORFs using the <code>write_orfs_fna</code> function. Similarly for the <code>BED</code> and <code>GFF</code> files using the <code>write_orfs_bed</code> and <code>write_orfs_gff</code> functions respectively.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../simplecodingrule/">« A Simple Coding Rule</a><a class="docs-footer-nextpage" href="../api/">API »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Wednesday 3 July 2024 15:49">Wednesday 3 July 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
