<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneFinder.jl</title><meta name="title" content="API · GeneFinder.jl"/><meta property="og:title" content="API · GeneFinder.jl"/><meta property="twitter:title" content="API · GeneFinder.jl"/><meta name="description" content="Documentation for GeneFinder.jl."/><meta property="og:description" content="Documentation for GeneFinder.jl."/><meta property="twitter:description" content="Documentation for GeneFinder.jl."/><meta property="og:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><meta property="twitter:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.svg" alt="GeneFinder.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../simplefinder/">Finding ORFs</a></li><li><a class="tocitem" href="../iodocs/">Wrtiting ORFs in files</a></li><li><a class="tocitem" href="../roadmap/">Roadmap</a></li><li class="is-active"><a class="tocitem" href>API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><ul><li><a href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a></li><li><a href="#GeneFinder.fasta_to_dna-Tuple{AbstractString}"><code>GeneFinder.fasta_to_dna</code></a></li><li><a href="#GeneFinder.findorfs-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.findorfs</code></a></li><li><a href="#GeneFinder.get_orfs_aa-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.get_orfs_aa</code></a></li><li><a href="#GeneFinder.get_orfs_dna-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.get_orfs_dna</code></a></li><li><a href="#GeneFinder.hasprematurestop-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.hasprematurestop</code></a></li><li><a href="#GeneFinder.locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.locationiterator</code></a></li><li><a href="#GeneFinder.record_orfs_faa-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.record_orfs_faa</code></a></li><li><a href="#GeneFinder.record_orfs_fna-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.record_orfs_fna</code></a></li><li><a href="#GeneFinder.write_orfs_bed-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N"><code>GeneFinder.write_orfs_bed</code></a></li><li><a href="#GeneFinder.write_orfs_faa-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N"><code>GeneFinder.write_orfs_faa</code></a></li><li><a href="#GeneFinder.write_orfs_fna-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N"><code>GeneFinder.write_orfs_fna</code></a></li><li><a href="#GeneFinder.write_orfs_gff-Union{Tuple{A}, Tuple{Union{BioSequences.LongSequence{A}, BioSequences.LongSubSeq{A}}, Union{IOStream, IOBuffer}}} where A"><code>GeneFinder.write_orfs_gff</code></a></li></ul><hr/><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.ORF" href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct ORF
    location::UnitRange{Int64}
    strand::Char
    frame::Int
end</code></pre><p>The ORF struct represents an open reading frame in a DNA sequence. It has two fields: </p><ul><li><code>location</code>: which is a UnitRange{Int64} indicating the start and end locations of the ORF in the sequence</li><li><code>strand</code>:  is a Char type indicating whether the ORF is on the forward (&#39;+&#39;) or reverse (&#39;-&#39;) strand of the sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.fasta_to_dna-Tuple{AbstractString}" href="#GeneFinder.fasta_to_dna-Tuple{AbstractString}"><code>GeneFinder.fasta_to_dna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fasta_to_dna(input::String)</code></pre><p>Converts a FASTA formatted file (even if it is a multi-fasta) to an array of <code>LongSequence{DNAAlphabet{4}}</code> objects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.findorfs-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.findorfs-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.findorfs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findorfs(sequence::NucleicAlphabet{DNAAlphabet{N}}; alternative_start::Bool=false, min_len::Int64=6)::Vector{ORF} where {N}</code></pre><p>A simple implementation that finds ORFs in a DNA sequence.</p><p>The <code>findorfs</code> function takes a LongSequence{DNAAlphabet{4}} sequence and returns a Vector{ORF} containing the ORFs found in the sequence.      It searches entire regularly expressed CDS, adding each ORF it finds to the vector. The function also searches the reverse complement of the sequence, so it finds ORFs on both strands.         Extending the starting codons with the <code>alternative_start = true</code> will search for ATG, GTG, and TTG.     Some studies have shown that in <em>E. coli</em> (K-12 strain), ATG, GTG and TTG are used 83 %, 14 % and 3 % respectively.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This function has not ORFs scoring scheme. Thus it might consider <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</p></div></div><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.get_orfs_aa-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.get_orfs_aa-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.get_orfs_aa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">get_orfs_aa(sequence::NucleicSeqOrView{DNAAlphabet{N}}; kwargs...) where {N}</code></pre><p>This function takes a <code>NucleicSeqOrView{DNAAlphabet{N}}</code> sequence and identifies the open reading frames (ORFs) using the <code>findorfs()</code> function. The function then translates each ORF into an amino acid sequence and stores it in a <code>Vector{LongSubSeq{AminoAcidAlphabet}}</code>.</p><p><strong>Arguments</strong></p><ul><li><code>sequence</code>: The input sequence as a <code>NucleicSeqOrView{DNAAlphabet{N}}</code></li></ul><p><strong>Keyword Arguments</strong></p><ul><li><code>alternative_start::Bool=false</code>: If set to <code>true</code>, the function considers alternative start codons when searching for ORFs. This increases the execution time by approximately 3x.</li><li><code>min_len::Int64=6</code>: The minimum length of the allowed ORF. By default, it allows ORFs that can encode at least one amino acid (e.g., <code>aa&quot;M*&quot;</code>).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.get_orfs_dna-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.get_orfs_dna-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.get_orfs_dna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">get_orfs_dna(sequence::NucleicSeqOrView{DNAAlphabet{N}}, output::String; kwargs...) where {N}</code></pre><p>This function takes a <code>NucleicSeqOrView{DNAAlphabet{N}}</code> sequence and identifies the open reading frames (ORFs) using the <code>findorfs()</code> function. The function then extracts the DNA sequence of each ORF and stores it in a <code>Vector{LongSubSeq{DNAAlphabet{4}}}</code>.</p><p><strong>Arguments</strong></p><ul><li><code>sequence</code>: The input sequence as a <code>NucleicSeqOrView{DNAAlphabet{N}}</code></li></ul><p><strong>Keyword Arguments</strong></p><ul><li><code>alternative_start::Bool=false</code>: If set to <code>true</code>, the function considers alternative start codons when searching for ORFs. This increases the execution time by approximately 3x.</li><li><code>min_len::Int64=6</code>: The minimum length of the allowed ORF. By default, it allows ORFs that can encode at least one amino acid (e.g., <code>aa&quot;M*&quot;</code>).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.hasprematurestop-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.hasprematurestop-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.hasprematurestop</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">hasprematurestop(sequence::LongNucOrView{4})::Bool</code></pre><p>Determine whether the <code>sequence</code> of type <code>LongSequence{DNAAlphabet{4}}</code> contains a premature stop codon.</p><p>Returns a boolean indicating whether the <code>sequence</code> has more than one stop codon.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.locationiterator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">locationiterator(sequence::NucleicSeqOrView{DNAAlphabet{N}}; alternative_start::Bool=false) where {N}</code></pre><p>This is an iterator function that uses regular expressions to search the entire ORF (instead of start and stop codons) in a <code>LongSequence{DNAAlphabet{4}}</code> sequence.     It uses an anonymous function that will find the first regularly expressed ORF. Then using this anonymous function it creates an iterator that will apply it until there is no other CDS.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>As a note of the implementation we want to expand on how the ORFs are found:</p><p>The expression <code>(?:[N]{3})*?</code> serves as the boundary between the start and stop codons. Within this expression, the character class <code>[N]{3}</code> captures exactly three occurrences of any character (representing nucleotides using IUPAC codes). This portion functions as the regular codon matches. Since it is enclosed within a non-capturing group <code>(?:)</code> and followed by <code>*?</code>, it allows for the matching of intermediate codons, but with a preference for the smallest number of repetitions. </p><p>In summary, the regular expression <code>ATG(?:[N]{3})*?T(AG|AA|GA)</code> identifies patterns that start with &quot;ATG,&quot; followed by any number of three-character codons (represented by &quot;N&quot; in the IUPAC code), and ends with a stop codon &quot;TAG,&quot; &quot;TAA,&quot; or &quot;TGA.&quot; This pattern is commonly used to identify potential protein-coding regions within genetic sequences.</p><p>See more about the discussion <a href="https://discourse.julialang.org/t/how-to-improve-a-generator-to-be-more-memory-efficient-when-it-is-collected/92932/8?u=camilogarciabotero">here</a></p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.record_orfs_faa-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.record_orfs_faa-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.record_orfs_faa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">record_orfs_faa(sequence::NucleicSeqOrView{DNAAlphabet{N}}; kwargs...) where {N}</code></pre><p>This function takes a nucleic acid sequence and finds all open reading frames (ORFs) in the sequence.  An ORF is a sequence of DNA that starts with a start codon and ends with a stop codon.  The function returns a list of FASTA records, where each record represents an ORF in the sequence.</p><p><strong>Arguments</strong></p><ul><li><code>sequence</code>: The nucleic acid sequence to search for ORFs.</li><li><code>alternative_start</code>: A boolean indicating whether to consider alternative start codons. Default is <code>false</code>.</li><li><code>code</code>: The genetic code to use for translation. Default is the NCBI translation table 1.</li><li><code>min_len</code>: The minimum length of an ORF. ORFs shorter than this length will be ignored. Default is 6.</li></ul><p><strong>Returns</strong></p><ul><li>A list of FASTA records representing the ORFs found in the sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.record_orfs_fna-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.record_orfs_fna-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.record_orfs_fna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">record_orfs_fna(sequence::NucleicSeqOrView{DNAAlphabet{N}}; kwargs...) where {N}</code></pre><p>Record Open Reading Frames (ORFs) in a nucleic acid sequence.</p><p><strong>Arguments</strong></p><ul><li><code>sequence::NucleicSeqOrView{DNAAlphabet{N}}</code>: The nucleic acid sequence to search for ORFs.</li><li><code>alternative_start::Bool=false</code>: If <code>true</code>, consider alternative start codons for ORFs.</li><li><code>min_len::Int=6</code>: The minimum length of an ORF to be recorded.</li></ul><p><strong>Returns</strong></p><p>An array of <code>FASTARecord</code> objects representing the identified ORFs.</p><p><strong>Description</strong></p><p>This function searches for Open Reading Frames (ORFs) in a given nucleic acid sequence. An ORF is a sequence of DNA that starts with a start codon and ends with a stop codon, without any other stop codons in between. By default, only the standard start codon (ATG) is considered, but if <code>alternative_start</code> is set to <code>true</code>, alternative start codons are also considered. The minimum length of an ORF to be recorded can be specified using the <code>min_len</code> argument.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_bed-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N" href="#GeneFinder.write_orfs_bed-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N"><code>GeneFinder.write_orfs_bed</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_bed(input::NucleicAcidAlphabet{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}; kwargs...) where {N}
write_orfs_bed(input::NucleicAcidAlphabet{DNAAlphabet{N}}, output::String; kwargs...) where {N}
write_orfs_bed(input::String, output::Union{IOStream, IOBuffer}; kwargs...)
write_orfs_bed(input::String, output::String; kwargs...)</code></pre><p>Write BED data to a file.</p><p><strong>Arguments</strong></p><ul><li><code>input::NucleicAcidAlphabet{DNAAlphabet{N}}</code>: The input DNA sequence.</li><li><code>output::String</code>: The output file path.</li><li><code>alternative_start::Bool=false</code>: If true, alternative start codons will be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len::Int64=6</code>: The minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_faa-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N" href="#GeneFinder.write_orfs_faa-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N"><code>GeneFinder.write_orfs_faa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_faa(input::LongSequence{DNAAlphabet{4}}, output::Union{IOStream, IOBuffer}; kwargs...) where {N}
write_orfs_faa(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...) where {N}
write_orfs_faa(input::String, output::Union{IOStream, IOBuffer}; kwargs...)
write_orfs_faa(input::String, output::String; kwargs...)</code></pre><p>Write the protein sequences encoded by the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">filename = &quot;output.faa&quot;

seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;

open(filename, &quot;w&quot;) do file
     write_orfs_faa(seq, file)
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_fna-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N" href="#GeneFinder.write_orfs_fna-Union{Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where N"><code>GeneFinder.write_orfs_fna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_fna(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}; kwargs...) where {N}
write_orfs_fna(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String; kwargs...) where {N}
write_orfs_fna(input::String, output::Union{IOStream, IOBuffer}; kwargs...)
write_orfs_fna(input::String, output::String; kwargs...)</code></pre><p>Write a file containing the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true, alternative start codons will be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len::Int64=6</code>: The minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">filename = &quot;output.fna&quot;

seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;

open(filename, &quot;w&quot;) do file
     write_orfs_fna(seq, file)
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_gff-Union{Tuple{A}, Tuple{Union{BioSequences.LongSequence{A}, BioSequences.LongSubSeq{A}}, Union{IOStream, IOBuffer}}} where A" href="#GeneFinder.write_orfs_gff-Union{Tuple{A}, Tuple{Union{BioSequences.LongSequence{A}, BioSequences.LongSubSeq{A}}, Union{IOStream, IOBuffer}}} where A"><code>GeneFinder.write_orfs_gff</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_gff(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}; kwargs...) where {N}
write_orfs_gff(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String; kwargs...) where {N}
write_orfs_gff(input::String, output::Union{IOStream, IOBuffer}; kwargs...)
write_orfs_gff(input::String, output::String; kwargs...)</code></pre><p>Write GFF data to a file.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: The input DNA sequence.</li><li><code>output</code>: The output file to write the GFF data to.</li><li><code>alternative_start::Bool=false</code>: If true, extended start codons will be considered during the search, increasing the execution time.</li><li><code>min_len::Int64=6</code>: The minimum length of the allowed ORF. The default value allows for possible encoding proteins with the <code>aa&quot;M*&quot;</code> sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../roadmap/">« Roadmap</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Tuesday 6 February 2024 01:49">Tuesday 6 February 2024</span>. Using Julia version 1.10.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
