<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneFinder.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../simplefinder/">A first algorithm</a></li><li class="is-active"><a class="tocitem" href>API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/main/docs/src/api.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.CDS" href="#GeneFinder.CDS"><code>GeneFinder.CDS</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct CDS
    orf::ORF
    sequence::LongSubSeq{DNAAlphabet{4}}
end</code></pre><p>The <code>CDS</code> struct represents a coding sequence in a DNA sequence. It has three fields:</p><ul><li><code>orf</code>: is the basic composible type (<code>location::UnitRange{Int}</code>, strand::Char) displaying the location of the ORF and the associate strand: forward (&#39;+&#39;) or reverse (&#39;-&#39;)</li><li><code>sequence</code>: a <code>LongDNA</code> sequence representing the actual sequence of the CDS</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/types.jl#L47-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.GeneFeatures" href="#GeneFinder.GeneFeatures"><code>GeneFinder.GeneFeatures</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct GeneFeatures
    seqname::String
    start::Int64
    stop::Int64
    score::Float64
    strand::Char
    frame::&#39;.&#39;
    attribute::
end</code></pre><p>This is the main Gene struct, based on the fields that could be found in a GFF3, still needs to be defined correctly,     The idea is correct the frame and attributes that will have something like a possible list (id=Char;name=;locus_tag).     The <code>write</code> and <code>get</code> functions should have a dedicated method for this struct.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/types.jl#L5-L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.ORF" href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct ORF
    location::UnitRange{Int64}
    strand::Char
end</code></pre><p>The ORF struct represents an open reading frame in a DNA sequence. It has two fields: </p><ul><li><code>location</code>: which is a UnitRange{Int64} indicating the start and end locations of the ORF in the sequence</li><li><code>strand</code>:  is a Char type indicating whether the ORF is on the forward (&#39;+&#39;) or reverse (&#39;-&#39;) strand of the sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/types.jl#L31-L41">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.Protein" href="#GeneFinder.Protein"><code>GeneFinder.Protein</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Protein
    sequence::LongSequence
    orf::ORF
end</code></pre><p>Similarly to the <code>CDS</code> struct, the <code>Protein</code> struct represents a encoded protein sequence in a DNA sequence.      It has three fields:</p><ul><li><code>orf</code>: is the basic composible type (<code>location::UnitRange{Int}</code>, strand::Char) of the sequence</li><li><code>sequence</code>: a <code>LongSequence</code> sequence representing the actual translated sequence of the CDS</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/types.jl#L63-L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.cdsgenerator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.cdsgenerator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.cdsgenerator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">cdsgenerator(sequence::LongDNA; kwargs...)
cdsgenerator(sequence::String; kwargs...)</code></pre><p>A function to generete CDSs sequence out of a DNA sequence.</p><p>The <code>cdsgenerator</code> is a generator function that takes a <code>LongDNA</code> sequence and returns an iterator over the given sequence,     containing the coding sequences (CDSs) found in the sequence and the ORF.      It uses the <code>findorfs</code> function to find open reading frames (ORFs) in the sequence,      and then it extracts the actual CDS sequence from each ORF, returining both.      The function also searches the reverse complement of the sequence, so it finds CDSs on both strands.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/findgenes.jl#L3-L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.eachcodon-Tuple{BioSequences.LongDNA}" href="#GeneFinder.eachcodon-Tuple{BioSequences.LongDNA}"><code>GeneFinder.eachcodon</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">eachcodon(sequence::LongDNA)</code></pre><p>Iterate through the codons in the <code>sequence</code> of type <code>LongDNA</code>.</p><p>Returns an iterator yielding <code>Codon</code> objects for each codon in the <code>sequence</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/helpers.jl#L45-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.fasta_to_dna-Tuple{String}" href="#GeneFinder.fasta_to_dna-Tuple{String}"><code>GeneFinder.fasta_to_dna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fasta_to_dna(input::String)</code></pre><p>Converts a FASTA formatted file (even if it is a multi-fasta) to an array of <code>LongDNA</code> objects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/helpers.jl#L34-L38">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.findorfs-Tuple{BioSequences.LongDNA}" href="#GeneFinder.findorfs-Tuple{BioSequences.LongDNA}"><code>GeneFinder.findorfs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findorfs(sequence::LongDNA; kwargs...)::Vector{ORF}
findorfs(sequence::String; kwargs...)::Vector{ORF}</code></pre><p>A simple implementation that finds ORFs in a DNA sequence.</p><p>The <code>findorfs</code> function takes a LongDNA sequence and returns a Vector{ORF} containing the ORFs found in the sequence.      It searches entire regularly expressed CDS, adding each ORF it finds to the vector. The function also searches the reverse complement of the sequence, so it finds ORFs on both strands.         Extending the starting codons with the <code>alternative_start = true</code> will search for ATG, GTG, and TTG.     Some studies have shown that in <em>E. coli</em> (K-12 strain), ATG, GTG and TTG are used 83 %, 14 % and 3 % respectively.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This function has not ORFs scoring scheme. Thus it might consider <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</p></div></div><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/algorithms/findorfs.jl#L14-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.getcds-Tuple{BioSequences.LongDNA}" href="#GeneFinder.getcds-Tuple{BioSequences.LongDNA}"><code>GeneFinder.getcds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getcds(input::LongDNA, output::String; kwargs...)
getcds(input::String, output::String; kwargs...)</code></pre><p>This function will take a <code>LongDNA</code> or <code>String</code> sequence and by means of the <code>findorfs()</code> function will push <code>LongSubSeq{DNAAlphabet{4}}</code> into a <code>Vector{}</code></p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/findgenes.jl#L48-L58">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.getproteins-Tuple{BioSequences.LongDNA}" href="#GeneFinder.getproteins-Tuple{BioSequences.LongDNA}"><code>GeneFinder.getproteins</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getproteins(input::LongDNA, output::String; kwargs...)
getproteins(input::String, output::String; kwargs...)</code></pre><p>Similar to <code>getcds()</code> function, it will take a <code>LongDNA</code> or <code>String</code> sequence and by means of the <code>findorfs()</code> and the <code>translate()</code> function will push <code>LongAA</code>s into a <code>Vector</code></p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/findgenes.jl#L118-L128">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.hasprematurestop-Tuple{BioSequences.LongDNA}" href="#GeneFinder.hasprematurestop-Tuple{BioSequences.LongDNA}"><code>GeneFinder.hasprematurestop</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">hasprematurestop(sequence::LongDNA)::Bool</code></pre><p>Determine whether the <code>sequence</code> of type <code>LongDNA</code> contains a premature stop codon.</p><p>Returns a boolean indicating whether the <code>sequence</code> has more than one stop codon.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/helpers.jl#L65-L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.locationiterator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.locationiterator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.locationiterator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">locationiterator(sequence::LongDNA; alternative_start::Bool=false)</code></pre><p>This is an iterator function that uses regular expressions to search the entire CDS (instead of start and stop codons) in a <code>LongDNA</code> sequence.     It uses an anonymous function that will find the first regularly expressed CDS. Then using this anonymous function it creates an iterator that will apply it until there is no other CDS.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/algorithms/findorfs.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.proteingenerator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.proteingenerator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.proteingenerator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">proteingenerator(sequence::LongDNA; kwargs...)
proteingenerator(sequence::String; kwargs...)</code></pre><p>As its name suggest this generator function iterates over the sequence to find proteins directly from a DNA sequence.      The <code>proteingenerator</code> function takes a <code>LongDNA</code> sequence and returns a <code>Vector{CDS}</code> containing the      coding sequences (CDSs) found in the sequence and the associated ORF.</p><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/findgenes.jl#L76-L89">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_bed-Tuple{BioSequences.LongDNA, String}" href="#GeneFinder.write_bed-Tuple{BioSequences.LongDNA, String}"><code>GeneFinder.write_bed</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_bed(input::LongDNA, output::String; kwargs...)
write_bed(input::String, output::String; kwargs...)</code></pre><p>Write BED data to a file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/io.jl#L1-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_cds-Tuple{BioSequences.LongDNA, String}" href="#GeneFinder.write_cds-Tuple{BioSequences.LongDNA, String}"><code>GeneFinder.write_cds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_cds(input::LongDNA, output::String; kwargs...)
write_cds(input::String, output::String; kwargs...)</code></pre><p>Write a file containing the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start</code>: A boolean value indicating whether alternative start codons should be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len</code>: An integer representing the minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/io.jl#L33-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_gff-Tuple{String, String}" href="#GeneFinder.write_gff-Tuple{String, String}"><code>GeneFinder.write_gff</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_gff(input::LongDNA, output::String; kwargs...)
write_gff(input::String, output::String; kwargs...)</code></pre><p>Write GFF data to a file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/io.jl#L97-L107">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_proteins-Tuple{BioSequences.LongDNA, String}" href="#GeneFinder.write_proteins-Tuple{BioSequences.LongDNA, String}"><code>GeneFinder.write_proteins</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_proteins(input::LongDNA, output::String; kwargs...)</code></pre><p>Write the protein sequences encoded by the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/c1fc7d40f88b24d794016b6f5ab8a088d85baf0a/src/io.jl#L65-L75">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../simplefinder/">« A first algorithm</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Thursday 8 June 2023 21:08">Thursday 8 June 2023</span>. Using Julia version 1.9.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
