<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneFinder.jl</title><meta name="title" content="API · GeneFinder.jl"/><meta property="og:title" content="API · GeneFinder.jl"/><meta property="twitter:title" content="API · GeneFinder.jl"/><meta name="description" content="Documentation for GeneFinder.jl."/><meta property="og:description" content="Documentation for GeneFinder.jl."/><meta property="twitter:description" content="Documentation for GeneFinder.jl."/><meta property="og:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><meta property="twitter:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.svg" alt="GeneFinder.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../naivefinder/">Finding ORFs</a></li><li><a class="tocitem" href="../features/">Scoring ORFs</a></li><li><a class="tocitem" href="../simplecodingrule/">A Simple Coding Rule</a></li><li><a class="tocitem" href="../iodocs/">Wrtiting ORFs In Files</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#The-Main-ORF-type"><span>The Main ORF type</span></a></li><li><a class="tocitem" href="#Finding-ORFs"><span>Finding ORFs</span></a></li><li><a class="tocitem" href="#Finding-ORFs-using-BioRegex-and-scoring"><span>Finding ORFs using BioRegex and scoring</span></a></li><li><a class="tocitem" href="#Writing-ORFs-to-files"><span>Writing ORFs to files</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h2 id="The-Main-ORF-type"><a class="docs-heading-anchor" href="#The-Main-ORF-type">The Main ORF type</a><a id="The-Main-ORF-type-1"></a><a class="docs-heading-anchor-permalink" href="#The-Main-ORF-type" title="Permalink"></a></h2><p>The main type of the package is <code>ORF</code> which represents an Open Reading Frame.</p><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.ORF" href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct ORF{N,F} &lt;: GenomicFeatures.AbstractGenomicInterval{F}</code></pre><p>The <code>ORF</code> struct represents an Open Reading Frame (ORF) in genomics.</p><p><strong>Fields</strong></p><ul><li><code>groupname::String</code>: The name of the group to which the ORF belongs.</li><li><code>first::Int64</code>: The starting position of the ORF.</li><li><code>last::Int64</code>: The ending position of the ORF.</li><li><code>strand::Strand</code>: The strand on which the ORF is located.</li><li><code>frame::Int</code>: The reading frame of the ORF.</li><li><code>features::Features</code>: The features associated with the ORF.</li><li><code>scheme::Union{Nothing,Function}</code>: The scheme used for the ORF.</li></ul><p><strong>Main Constructor</strong></p><pre><code class="language-julia hljs">ORF{N,F}(
    groupname::String,
    first::Int64,
    last::Int64,
    strand::Strand,
    frame::Int,
    features::Features,
    scheme::Union{Nothing,Function}
)</code></pre><p><strong>Example</strong></p><p>A full instance <code>ORF</code></p><pre><code class="language-julia hljs">ORF{4,NaiveFinder}(&quot;seq01&quot;, 1, 33, STRAND_POS, 1, Features((score = 0.0,)), nothing)</code></pre><p>A partial instance <code>ORF</code></p><pre><code class="language-julia hljs">ORF{NaiveFinder}(1:33, &#39;+&#39;, 1)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.features-Union{Tuple{ORF{N, F}}, Tuple{F}, Tuple{N}} where {N, F}" href="#GeneFinder.features-Union{Tuple{ORF{N, F}}, Tuple{F}, Tuple{N}} where {N, F}"><code>GeneFinder.features</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">features(i::ORF{N,F})</code></pre><p>Extracts the features from an <code>ORF</code> object.</p><p><strong>Arguments</strong></p><ul><li><code>i::ORF{N,F}</code>: An <code>ORF</code> object.</li></ul><p><strong>Returns</strong></p><p>The features of the <code>ORF</code> object. Those could be defined by each <code>GeneFinderMethod</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.sequence-Union{Tuple{ORF{N, F}}, Tuple{F}, Tuple{N}} where {N, F}" href="#GeneFinder.sequence-Union{Tuple{ORF{N, F}}, Tuple{F}, Tuple{N}} where {N, F}"><code>GeneFinder.sequence</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence(i::ORF{N,F})</code></pre><p>Extracts the DNA sequence corresponding to the given open reading frame (ORF).</p><p><strong>Arguments</strong></p><ul><li><code>i::ORF{N,F}</code>: The open reading frame (ORF) for which the DNA sequence needs to be extracted.</li></ul><p><strong>Returns</strong></p><ul><li>The DNA sequence corresponding to the given open reading frame (ORF).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.source-Union{Tuple{ORF{N, F}}, Tuple{F}, Tuple{N}} where {N, F}" href="#GeneFinder.source-Union{Tuple{ORF{N, F}}, Tuple{F}, Tuple{N}} where {N, F}"><code>GeneFinder.source</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">source(i::ORF{N,F})</code></pre><p>Get the source sequence associated with the given <code>ORF</code> object.</p><p><strong>Arguments</strong></p><ul><li><code>i::ORF{N,F}</code>: The <code>ORF</code> object for which to retrieve the source sequence.</li></ul><p><strong>Returns</strong></p><p>The source sequence associated with the <code>ORF</code> object.</p><p><strong>Examples</strong></p><pre><code class="language-julia hljs">seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;
orfs = findorfs(seq)
source(orfs[1])

44nt DNA Sequence:
ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA</code></pre><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>The <code>source</code> method works if the sequence is defined in the global scope. Otherwise it will throw an error.  For instance a common failure is to define a simple <code>ORF</code> that by defualt will have an &quot;unnamedsource&quot; as <code>groupname</code>  and then try to get the source sequence. </p><pre><code class="language-julia hljs">orf = ORF{NaiveFinder}(1:33, &#39;+&#39;, 1)
source(orf)

ERROR: UndefVarError: `unnamedsource` not defined
Stacktrace:
 [1] source(i::ORF{4, NaiveFinder})
   @ GeneFinder ~/.julia/dev/GeneFinder/src/types.jl:192
 [2] top-level scope
   @ REPL[12]:1</code></pre></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Finding-ORFs"><a class="docs-heading-anchor" href="#Finding-ORFs">Finding ORFs</a><a id="Finding-ORFs-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-ORFs" title="Permalink"></a></h2><p>The function <code>findorfs</code> is the main function of the package. It is generic method that can handle different gene finding methods. </p><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.findorfs-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{F}, Tuple{N}} where {N, F&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.findorfs-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{F}, Tuple{N}} where {N, F&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.findorfs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findorfs(sequence::NucleicSeqOrView{DNAAlphabet{N}}; ::F, kwargs...) where {N, F&lt;:GeneFinderMethod}</code></pre><p>This is the main interface method for finding open reading frames (ORFs) in a DNA sequence.</p><p>It takes the following required arguments:</p><ul><li><code>sequence</code>: The nucleic acid sequence to search for ORFs.</li><li><code>finder</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder</code>, <code>NaiveCollector</code> or yet other implementations.</li></ul><p><strong>Keyword Arguments regardless of the finder method:</strong></p><ul><li><code>alternative_start::Bool</code>: A boolean indicating whether to consider alternative start codons. Default is <code>false</code>.</li><li><code>minlen::Int</code>: The minimum length of an ORF. Default is <code>6</code>.</li><li><code>scheme::Function</code>: The scoring scheme to use for scoring the sequence from the ORF. Default is <code>nothing</code>.</li></ul><p><strong>Returns</strong></p><p>A vector of <code>ORF</code> objects representing the found ORFs.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">sequence = randdnaseq(120)

120nt DNA Sequence:
 GCCGGACAGCGAAGGCTAATAAATGCCCGTGCCAGTATC…TCTGAGTTACTGTACACCCGAAAGACGTTGTACGCATTT

findorfs(sequence, finder=NaiveFinder)

1-element Vector{ORF}:
 ORF{NaiveFinder}(77:118, &#39;-&#39;, 2)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Finding-ORFs-using-BioRegex-and-scoring"><a class="docs-heading-anchor" href="#Finding-ORFs-using-BioRegex-and-scoring">Finding ORFs using BioRegex and scoring</a><a id="Finding-ORFs-using-BioRegex-and-scoring-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-ORFs-using-BioRegex-and-scoring" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.NaiveFinder-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.NaiveFinder-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.NaiveFinder</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">NaiveFinder(sequence::NucleicSeqOrView{DNAAlphabet{N}}; kwargs...) -&gt; Vector{ORF} where {N}</code></pre><p>A simple implementation that finds ORFs in a DNA sequence.</p><p>The <code>NaiveFinder</code> method takes a LongSequence{DNAAlphabet{4}} sequence and returns a Vector{ORF} containing the ORFs found in the sequence.      It searches entire regularly expressed CDS, adding each ORF it finds to the vector. The function also searches the reverse complement of the sequence, so it finds ORFs on both strands.         Extending the starting codons with the <code>alternative_start = true</code> will search for ATG, GTG, and TTG.     Some studies have shown that in <em>E. coli</em> (K-12 strain), ATG, GTG and TTG are used 83 %, 14 % and 3 % respectively.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This function has neither ORFs scoring scheme by default nor length constraints. Thus it might consider <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</p></div></div><p><strong>Required Arguments</strong></p><ul><li><code>sequence::NucleicSeqOrView{DNAAlphabet{N}}</code>: The nucleic acid sequence to search for ORFs.</li></ul><p><strong>Keywords Arguments</strong></p><ul><li><code>alternative_start::Bool</code>: If true will pass the extended start codons to search. This will increase 3x the execution time. Default is <code>false</code>.</li><li><code>minlen::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li><li><code>scheme::Function</code>: The scoring scheme to use for scoring the sequence from the ORF. Default is <code>nothing</code>.</li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>As the scheme is generally a scoring function that at least requires a sequence, one simple scheme is the log-odds ratio score. This score is a log-odds ratio that compares the probability of the sequence generated by a coding model to the probability of the sequence generated by a non-coding model:</p><p class="math-container">\[S(x) = \sum_{i=1}^{L} \beta_{x_{i}x} = \sum_{i=1} \log \frac{a^{\mathscr{m}_{1}}_{i-1} x_i}{a^{\mathscr{m}_{2}}_{i-1} x_i}\]</p><p>If the log-odds ratio exceeds a given threshold (<code>η</code>), the sequence is considered likely to be coding. See <a href="@ref"><code>lordr</code></a> for more information about coding creteria.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder._locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder._locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder._locationiterator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">_locationiterator(sequence::NucleicSeqOrView{DNAAlphabet{N}}; alternative_start::Bool=false) where {N}</code></pre><p>This is an iterator function that uses regular expressions to search the entire ORF (instead of start and stop codons) in a <code>LongSequence{DNAAlphabet{4}}</code> sequence.     It uses an anonymous function that will find the first regularly expressed ORF. Then using this anonymous function it creates an iterator that will apply it until there is no other CDS.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>As a note of the implementation we want to expand on how the ORFs are found:</p><p>The expression <code>(?:[N]{3})*?</code> serves as the boundary between the start and stop codons. Within this expression, the character class <code>[N]{3}</code> captures exactly three occurrences of any character (representing nucleotides using IUPAC codes). This portion functions as the regular codon matches. Since it is enclosed within a non-capturing group <code>(?:)</code> and followed by <code>*?</code>, it allows for the matching of intermediate codons, but with a preference for the smallest number of repetitions. </p><p>In summary, the regular expression <code>ATG(?:[N]{3})*?T(AG|AA|GA)</code> identifies patterns that start with &quot;ATG,&quot; followed by any number of three-character codons (represented by &quot;N&quot; in the IUPAC code), and ends with a stop codon &quot;TAG,&quot; &quot;TAA,&quot; or &quot;TGA.&quot; This pattern is commonly used to identify potential protein-coding regions within genetic sequences.</p><p>See more about the discussion <a href="https://discourse.julialang.org/t/how-to-improve-a-generator-to-be-more-memory-efficient-when-it-is-collected/92932/8?u=camilogarciabotero">here</a></p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Writing-ORFs-to-files"><a class="docs-heading-anchor" href="#Writing-ORFs-to-files">Writing ORFs to files</a><a id="Writing-ORFs-to-files-1"></a><a class="docs-heading-anchor-permalink" href="#Writing-ORFs-to-files" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_bed-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_bed-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_bed</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_bed(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}, finder::F; kwargs...)
write_orfs_bed(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String, finder::F; kwargs...)</code></pre><p>Write BED data to a file.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: The input DNA sequence NucSeq or a view.</li><li><code>output</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>finder</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true, alternative start codons will be used when identifying CDSs. Default is <code>false</code>.</li><li><code>minlen::Int64=6</code>: The minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_faa-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_faa-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_faa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_faa(input::NucleicSeqOrView{DNAAlphabet{4}}, output::Union{IOStream, IOBuffer}, finder::F; kwargs...)
write_orfs_faa(input::NucleicSeqOrView{DNAAlphabet{4}}, output::String, finder::F; kwargs...)</code></pre><p>Write the protein sequences encoded by the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: The input DNA sequence NucSeq or a view.</li><li><code>output</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>finder</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>minlen::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">filename = &quot;output.faa&quot;

seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;

open(filename, &quot;w&quot;) do file
     write_orfs_faa(seq, file)
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_fna-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_fna-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_fna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_fna(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}, finder::F; kwargs...)
write_orfs_fna(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String, finder::F; kwargs...)</code></pre><p>Write a file containing the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Arguments</strong></p><ul><li><code>input::NucleicAcidAlphabet{DNAAlphabet{N}}</code>: The input DNA sequence.</li><li><code>output::IO</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>finder::F</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true, alternative start codons will be used when identifying CDSs. Default is <code>false</code>.</li><li><code>minlen::Int64=6</code>: The minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">filename = &quot;output.fna&quot;

seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;

open(filename, &quot;w&quot;) do file
     write_orfs_fna(seq, file, NaiveFinder())
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_gff-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_gff-Union{Tuple{F}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}}} where {N, F&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_gff</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_gff(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}, finder::F; kwargs...)
write_orfs_gff(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String, finder::F; kwargs...)</code></pre><p>Write GFF data to a file.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: The input DNA sequence NucSeq or a view.</li><li><code>output</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>finder</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>minlen::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../iodocs/">« Wrtiting ORFs In Files</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Monday 8 July 2024 16:56">Monday 8 July 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
