<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneFinder.jl</title><meta name="title" content="API · GeneFinder.jl"/><meta property="og:title" content="API · GeneFinder.jl"/><meta property="twitter:title" content="API · GeneFinder.jl"/><meta name="description" content="Documentation for GeneFinder.jl."/><meta property="og:description" content="Documentation for GeneFinder.jl."/><meta property="twitter:description" content="Documentation for GeneFinder.jl."/><meta property="og:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><meta property="twitter:url" content="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.svg" alt="GeneFinder.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../naivefinder/">Finding ORFs</a></li><li><a class="tocitem" href="../simplecodingrule/">A Simple Coding Rule</a></li><li><a class="tocitem" href="../iodocs/">Wrtiting ORFs In Files</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#The-Main-ORF-type"><span>The Main ORF type</span></a></li><li><a class="tocitem" href="#Finding-ORFs"><span>Finding ORFs</span></a></li><li><a class="tocitem" href="#Finding-ORFs-using-BioRegex-and-scoring"><span>Finding ORFs using BioRegex and scoring</span></a></li><li><a class="tocitem" href="#Geting-ORFs-sequences"><span>Geting ORFs sequences</span></a></li><li><a class="tocitem" href="#Writing-ORFs-to-files"><span>Writing ORFs to files</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h2 id="The-Main-ORF-type"><a class="docs-heading-anchor" href="#The-Main-ORF-type">The Main ORF type</a><a id="The-Main-ORF-type-1"></a><a class="docs-heading-anchor-permalink" href="#The-Main-ORF-type" title="Permalink"></a></h2><p>The main type of the package is <code>ORF</code> which represents an Open Reading Frame.</p><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.ORF" href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct ORF
    location::UnitRange{Int64}
    strand::Char
    frame::Int
    score::Union{Nothing, Float64}
end</code></pre><p>The ORF struct represents an open reading frame in a DNA sequence. It has two fields: </p><ul><li><code>location</code>: which is a UnitRange{Int64} indicating the start and end locations of the ORF in the sequence</li><li><code>strand</code>:  is a Char type indicating whether the ORF is on the forward (&#39;+&#39;) or reverse (&#39;-&#39;) strand of the sequence.</li><li><code>frame</code>: is an Int type indicating the reading frame of the ORF. The frame is the position of the first nucleotide of the codon that starts the ORF, relative to the start of the sequence. It can be 1, 2, or 3.</li><li><code>score</code>: is a Union{Nothing, Float64} type indicating the score of the ORF. It can be a Float64 or nothing.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Finding-ORFs"><a class="docs-heading-anchor" href="#Finding-ORFs">Finding ORFs</a><a id="Finding-ORFs-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-ORFs" title="Permalink"></a></h2><p>The function <code>findorfs</code> is the main function of the package. It is generic method that can handle different gene finding methods. </p><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.findorfs" href="#GeneFinder.findorfs"><code>GeneFinder.findorfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">findorfs(sequence::NucleicSeqOrView{DNAAlphabet{N}}; ::M, kwargs...) where {N, M&lt;:GeneFinderMethod}</code></pre><p>This is the main interface method for finding open reading frames (ORFs) in a DNA sequence.</p><p>It takes the following arguments:</p><ul><li><code>sequence</code>: The nucleic acid sequence to search for ORFs.</li><li><code>method</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code>, <code>NaiveFinderScored()</code> or yet other implementations.</li></ul><p><strong>Keyword Arguments regardless of the finder method:</strong></p><ul><li><code>alternative_start</code>: A boolean indicating whether to consider alternative start codons. Default is <code>false</code>.</li><li><code>min_len</code>: The minimum length of an ORF. Default is <code>6</code>.</li></ul><p><strong>Keyword Arguments for <code>NaiveFinderScored()</code>:</strong></p><ul><li><code>scoringscheme::BioMarkovChain</code>: The scoring scheme to use for the scoring algorithm. Default is <code>ECOLICDS</code>.</li></ul><p><strong>Returns</strong></p><p>A vector of <code>ORF</code> objects representing the found ORFs.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">sequence = randdnaseq(120)

    120nt DNA Sequence:
    GCCGGACAGCGAAGGCTAATAAATGCCCGTGCCAGTATC…TCTGAGTTACTGTACACCCGAAAGACGTTGTACGCATTT


findorfs(sequence, NaiveFinder())

    1-element Vector{ORF}:
    ORF(77:118, &#39;-&#39;, 2, 0.0)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Finding-ORFs-using-BioRegex-and-scoring"><a class="docs-heading-anchor" href="#Finding-ORFs-using-BioRegex-and-scoring">Finding ORFs using BioRegex and scoring</a><a id="Finding-ORFs-using-BioRegex-and-scoring-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-ORFs-using-BioRegex-and-scoring" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder._locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder._locationiterator-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder._locationiterator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">locationiterator(sequence::NucleicSeqOrView{DNAAlphabet{N}}; alternative_start::Bool=false) where {N}</code></pre><p>This is an iterator function that uses regular expressions to search the entire ORF (instead of start and stop codons) in a <code>LongSequence{DNAAlphabet{4}}</code> sequence.     It uses an anonymous function that will find the first regularly expressed ORF. Then using this anonymous function it creates an iterator that will apply it until there is no other CDS.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>As a note of the implementation we want to expand on how the ORFs are found:</p><p>The expression <code>(?:[N]{3})*?</code> serves as the boundary between the start and stop codons. Within this expression, the character class <code>[N]{3}</code> captures exactly three occurrences of any character (representing nucleotides using IUPAC codes). This portion functions as the regular codon matches. Since it is enclosed within a non-capturing group <code>(?:)</code> and followed by <code>*?</code>, it allows for the matching of intermediate codons, but with a preference for the smallest number of repetitions. </p><p>In summary, the regular expression <code>ATG(?:[N]{3})*?T(AG|AA|GA)</code> identifies patterns that start with &quot;ATG,&quot; followed by any number of three-character codons (represented by &quot;N&quot; in the IUPAC code), and ends with a stop codon &quot;TAG,&quot; &quot;TAA,&quot; or &quot;TGA.&quot; This pattern is commonly used to identify potential protein-coding regions within genetic sequences.</p><p>See more about the discussion <a href="https://discourse.julialang.org/t/how-to-improve-a-generator-to-be-more-memory-efficient-when-it-is-collected/92932/8?u=camilogarciabotero">here</a></p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.log_odds_ratio_decision_rule-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.log_odds_ratio_decision_rule-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.log_odds_ratio_decision_rule</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">log_odds_ratio_decision_rule(
    sequence::LongSequence{DNAAlphabet{4}};
    codingmodel::BioMarkovChain,
    noncodingmodel::BioMarkovChain,
    η::Float64 = 1e-5
    )</code></pre><p>Check if a given DNA sequence is likely to be coding based on a log-odds ratio.     The log-odds ratio is a statistical measure used to assess the likelihood of a sequence being coding or non-coding. It compares the probability of the sequence generated by a coding model to the probability of the sequence generated by a non-coding model. If the log-odds ratio exceeds a given threshold (<code>η</code>), the sequence is considered likely to be coding.     It is formally described as a decision rule:</p><p class="math-container">\[S(X) = \log \left( \frac{{P_C(X_1=i_1, \ldots, X_T=i_T)}}{{P_N(X_1=i_1, \ldots, X_T=i_T)}} \right) \begin{cases} &gt; \eta &amp; \Rightarrow \text{{coding}} \\ &lt; \eta &amp; \Rightarrow \text{{noncoding}} \end{cases}\]</p><p><strong>Arguments</strong></p><ul><li><code>sequence::NucleicSeqOrView{DNAAlphabet{N}}</code>: The DNA sequence to be evaluated.</li></ul><p><strong>Keyword Arguments</strong></p><ul><li><code>codingmodel::BioMarkovChain</code>: The transition model for coding regions, (default: <code>ECOLICDS</code>).</li><li><code>noncodingmodel::BioMarkovChain</code>: The transition model for non-coding regions, (default: <code>ECOLINOCDS</code>)</li><li><code>η::Float64 = 1e-5</code>: The threshold value (eta) for the log-odds ratio (default: 1e-5).</li></ul><p><strong>Returns</strong></p><ul><li><code>true</code> if the sequence is likely to be coding.</li><li><code>false</code> if the sequence is likely to be non-coding.</li></ul><p><strong>Raises</strong></p><ul><li><code>ErrorException</code>: if the length of the sequence is not divisible by 3.</li><li><code>ErrorException</code>: if the sequence contains a premature stop codon.</li></ul><p><strong>Example</strong></p><pre><code class="nohighlight hljs">sequence = dna&quot;ATGGCATCTAG&quot;
iscoding(sequence)  # Returns: true or false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.naivefinder-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.naivefinder-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.naivefinder</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">naivefinder(sequence::NucleicAlphabet{DNAAlphabet{N}}; alternative_start::Bool=false, min_len::Int64=6)::Vector{ORF} where {N}</code></pre><p>A simple implementation that finds ORFs in a DNA sequence.</p><p>The <code>naivefinder</code> function takes a LongSequence{DNAAlphabet{4}} sequence and returns a Vector{ORF} containing the ORFs found in the sequence.      It searches entire regularly expressed CDS, adding each ORF it finds to the vector. The function also searches the reverse complement of the sequence, so it finds ORFs on both strands.         Extending the starting codons with the <code>alternative_start = true</code> will search for ATG, GTG, and TTG.     Some studies have shown that in <em>E. coli</em> (K-12 strain), ATG, GTG and TTG are used 83 %, 14 % and 3 % respectively.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This function has not ORFs scoring scheme. Thus it might consider <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</p></div></div><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.naivefinderscored-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N" href="#GeneFinder.naivefinderscored-Union{Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}}, Tuple{N}} where N"><code>GeneFinder.naivefinderscored</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">naivefinderscored(sequence::NucleicSeqOrView{DNAAlphabet{N}}; alternative_start=false, min_len=6) where {N}</code></pre><p>Find open reading frames (ORFs) in a nucleic acid sequence using a naive scoring algorithm. In this method the scoring is a log-odds ratio score defines as:</p><p class="math-container">\[S(x) = \sum_{i=1}^{L} \beta_{x_{i}x} = \sum_{i=1} \log \frac{a^{\mathscr{m}_{1}}_{i-1} x_i}{a^{\mathscr{m}_{2}}_{i-1} x_i}\]</p><p><strong>Arguments</strong></p><ul><li><code>sequence</code>: The nucleic acid sequence to search for ORFs.</li><li><code>alternative_start</code>: A boolean indicating whether to consider alternative start codons. Default is <code>false</code>.</li><li><code>min_len</code>: The minimum length of an ORF to be considered. Default is <code>6</code>.</li><li><code>scoringscheme</code>: The scoring scheme to use for the scoring algorithm. Default is <code>ECOLICDS</code>.</li></ul><p><strong>Returns</strong></p><p>A sorted vector of <code>ORF</code> objects representing the identified open reading frames.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Geting-ORFs-sequences"><a class="docs-heading-anchor" href="#Geting-ORFs-sequences">Geting ORFs sequences</a><a id="Geting-ORFs-sequences-1"></a><a class="docs-heading-anchor-permalink" href="#Geting-ORFs-sequences" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.getorfs-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, BioSequences.DNAAlphabet{N}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.getorfs-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, BioSequences.DNAAlphabet{N}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.getorfs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getorfs(sequence::NucleicSeqOrView{DNAAlphabet{N}}, ::Alphabet, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}</code></pre><p>This function takes a <code>NucleicSeqOrView{DNAAlphabet{N}}</code> sequence and identifies the open reading frames (ORFs) using the <code>findorfs()</code> function under the selected finder method. The function then extracts the DNA or aminoacid sequence of each ORF and stores it in a <code>Vector{LongSubSeq{A}}</code>.</p><p><strong>Arguments</strong></p><ul><li><code>sequence</code>: The input sequence as a <code>NucleicSeqOrView{DNAAlphabet{N}}</code></li><li><code>alphabet</code>: The output sequence alphabet. It can be either <code>DNAAlphabet{4}()</code> or <code>AminoAcidAlphabet()</code>.</li><li><code>method</code>: The method used to find ORFs. It must be a subtype of <code>GeneFinderMethod</code>. (e.g., <code>NaiveFinder()</code>)</li></ul><p><strong>Keyword Arguments</strong></p><ul><li><code>alternative_start::Bool=false</code>: If set to <code>true</code>, the function considers alternative start codons when searching for ORFs. This increases the execution time by approximately 3x.</li><li><code>min_len::Int64=6</code>: The minimum length of the allowed ORF. By default, it allows ORFs that can encode at least one amino acid (e.g., <code>aa&quot;M*&quot;</code>).</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">sequence = randdnaseq(120)

    120nt DNA Sequence:
    GCCGGACAGCGAAGGCTAATAAATGCCCGTGCCAGTATC…TCTGAGTTACTGTACACCCGAAAGACGTTGTACGCATTT


getorfs(sequence, DNAAlphabet{4}(), NaiveFinder())

    1-element Vector{LongSubSeq{DNAAlphabet{4}}}:
     ATGCGTACAACGTCTTTCGGGTGTACAGTAACTCAGAGCTAA
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><h2 id="Writing-ORFs-to-files"><a class="docs-heading-anchor" href="#Writing-ORFs-to-files">Writing ORFs to files</a><a id="Writing-ORFs-to-files-1"></a><a class="docs-heading-anchor-permalink" href="#Writing-ORFs-to-files" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_bed-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_bed-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_bed</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_bed(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_bed(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String, ::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_bed(input::String, output::Union{IOStream, IOBuffer}, method::M; kwargs...) where {M&lt;:GeneFinderMethod}
write_orfs_bed(input::String, output::String, method::M; kwargs...) where {M&lt;:GeneFinderMethod}</code></pre><p>Write BED data to a file.</p><p><strong>Arguments</strong></p><ul><li><code>input::NucleicAcidAlphabet{DNAAlphabet{N}}</code>: The input DNA sequence.</li><li><code>output::IO</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>method::M</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true, alternative start codons will be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len::Int64=6</code>: The minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_faa-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_faa-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_faa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_faa(input::NucleicSeqOrView{DNAAlphabet{4}}, output::Union{IOStream, IOBuffer}, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_faa(input::NucleicSeqOrView{DNAAlphabet{4}}, output::String, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_faa(input::String, output::Union{IOStream, IOBuffer}, method::M; kwargs...) where {M&lt;:GeneFinderMethod}
write_orfs_faa(input::String, output::String, method::M; kwargs...) where {M&lt;:GeneFinderMethod}</code></pre><p>Write the protein sequences encoded by the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Arguments</strong></p><ul><li><code>input::NucleicAcidAlphabet{DNAAlphabet{N}}</code>: The input DNA sequence.</li><li><code>output::IO</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>method::M</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">filename = &quot;output.faa&quot;

seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;

open(filename, &quot;w&quot;) do file
     write_orfs_faa(seq, file)
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_fna-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_fna-Union{Tuple{M}, Tuple{N}, Tuple{Union{BioSequences.LongDNA{N}, BioSequences.LongSubSeq{BioSequences.DNAAlphabet{N}}}, Union{IOStream, IOBuffer}, M}} where {N, M&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_fna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_fna(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_fna(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_fna(input::String, output::Union{IOStream, IOBuffer}; kwargs...) where {M&lt;:GeneFinderMethod}
write_orfs_fna(input::String, output::String, method::M; kwargs...) where {M&lt;:GeneFinderMethod}</code></pre><p>Write a file containing the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Arguments</strong></p><ul><li><code>input::NucleicAcidAlphabet{DNAAlphabet{N}}</code>: The input DNA sequence.</li><li><code>output::IO</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>method::M</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true, alternative start codons will be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len::Int64=6</code>: The minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">filename = &quot;output.fna&quot;

seq = dna&quot;ATGATGCATGCATGCATGCTAGTAACTAGCTAGCTAGCTAGTAA&quot;

open(filename, &quot;w&quot;) do file
     write_orfs_fna(seq, file)
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneFinder.write_orfs_gff-Union{Tuple{M}, Tuple{A}, Tuple{Union{BioSequences.LongSequence{A}, BioSequences.LongSubSeq{A}}, Union{IOStream, IOBuffer}, M}} where {A, M&lt;:GeneFinder.GeneFinderMethod}" href="#GeneFinder.write_orfs_gff-Union{Tuple{M}, Tuple{A}, Tuple{Union{BioSequences.LongSequence{A}, BioSequences.LongSubSeq{A}}, Union{IOStream, IOBuffer}, M}} where {A, M&lt;:GeneFinder.GeneFinderMethod}"><code>GeneFinder.write_orfs_gff</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_orfs_gff(input::NucleicSeqOrView{DNAAlphabet{N}}, output::Union{IOStream, IOBuffer}, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_gff(input::NucleicSeqOrView{DNAAlphabet{N}}, output::String, method::M; kwargs...) where {N, M&lt;:GeneFinderMethod}
write_orfs_gff(input::String, output::Union{IOStream, IOBuffer}, method::M; kwargs...)  where {M&lt;:GeneFinderMethod}
write_orfs_gff(input::String, output::String, method::M; kwargs...) where {M&lt;:GeneFinderMethod}</code></pre><p>Write GFF data to a file.</p><p><strong>Arguments</strong></p><ul><li><code>input::NucleicAcidAlphabet{DNAAlphabet{N}}</code>: The input DNA sequence.</li><li><code>output::IO</code>: The otput format, it can be a file (<code>String</code>) or a buffer (<code>IOStream</code> or `IOBuffer)</li><li><code>method::M</code>: The algorithm used to find ORFs. It can be either <code>NaiveFinder()</code> or <code>NaiveFinderScored()</code>.</li></ul><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../iodocs/">« Wrtiting ORFs In Files</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Tuesday 14 May 2024 06:15">Tuesday 14 May 2024</span>. Using Julia version 1.10.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
