<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneFinder.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../simplefinder/">A first algorithm</a></li><li class="is-active"><a class="tocitem" href>API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/main/docs/src/api.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.CDS" href="#GeneFinder.CDS"><code>GeneFinder.CDS</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct CDS
    orf::ORF
    sequence::LongDNA
end</code></pre><p>The <code>CDS</code> struct represents a coding sequence in a DNA sequence. It has three fields:</p><ul><li><code>orf</code>: is the basic composible type (<code>location::UnitRange{Int}</code>, strand::Char) displaying the location of the ORF and the associate strand: forward (&#39;+&#39;) or reverse (&#39;-&#39;)</li><li><code>sequence</code>: a <code>LongDNA</code> sequence representing the actual sequence of the CDS</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/types.jl#L22-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.Codon" href="#GeneFinder.Codon"><code>GeneFinder.Codon</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Codon &lt;: BioSequence{DNAAlphabet{2}}</code></pre><p>A <code>Struct</code> representing a codon, which is a subtype of <code>BioSequence</code> with an <code>Alphabet</code> of type <code>DNAAlphabet{2}</code>. It has a single field <code>x</code> of type <code>UInt8</code>. This was implemente in The following implementation is from https://biojulia.net/BioSequences.jl/stable/interfaces/</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/types.jl#L56-L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.ORF" href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct ORF
    location::UnitRange{Int64}
    strand::Char
    
end</code></pre><p>The ORF struct represents an open reading frame in a DNA sequence. It has two fields: </p><ul><li><code>location</code>: which is a UnitRange{Int64} indicating the start and end locations of the ORF in the sequence</li><li><code>strand</code>:  is a Char type indicating whether the ORF is on the forward (&#39;+&#39;) or reverse (&#39;-&#39;) strand of the sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/types.jl#L5-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.Protein" href="#GeneFinder.Protein"><code>GeneFinder.Protein</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Protein
    sequence::LongSequence
    orf::ORF
end</code></pre><p>Similarly to the <code>CDS</code> struct, the <code>Protein</code> struct represents a encoded protein sequence in a DNA sequence.      It has three fields:</p><ul><li><code>orf</code>: is the basic composible type (<code>location::UnitRange{Int}</code>, strand::Char) of the sequence</li><li><code>sequence</code>: a <code>LongSequence</code> sequence representing the actual translated sequence of the CDS</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/types.jl#L38-L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.eachcodon-Tuple{BioSequences.LongDNA}" href="#GeneFinder.eachcodon-Tuple{BioSequences.LongDNA}"><code>GeneFinder.eachcodon</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">eachcodon(sequence::LongDNA)</code></pre><p>Iterate through the codons in the <code>sequence</code> of type <code>LongDNA</code>.</p><p>Returns an iterator yielding <code>Codon</code> objects for each codon in the <code>sequence</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/helpers.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.hasprematurestop-Tuple{BioSequences.LongDNA}" href="#GeneFinder.hasprematurestop-Tuple{BioSequences.LongDNA}"><code>GeneFinder.hasprematurestop</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">hasprematurestop(sequence::LongDNA)::Bool</code></pre><p>Determine whether the <code>sequence</code> of type <code>LongDNA</code> contains a premature stop codon.</p><p>Returns a boolean indicating whether the <code>sequence</code> has more than one stop codon.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/helpers.jl#L21-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.locationgenerator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.locationgenerator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.locationgenerator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">locationgenerator(sequence::LongDNA; alternative_start::Bool=false)</code></pre><p>Generate the locations of ORFs in the given DNA <code>sequence</code>.</p><p>This function searches the sequence for start codons, and generates ranges of indices corresponding to the locations of ORFs in the <code>sequence</code>.      The ORFs are generated by iterating over the start codon indices and searching for the first stop codon that follows each start codon.          ORFs that contain premature stop codons are filtered out using the <code>hasprematurestop</code> function.              The <code>sequence</code> argument must be a <code>LongDNA</code> object, which is a type of DNA sequence with a longer maximum length than the <code>DNA</code> type.</p><p>Returns:     A generator expression that yields ranges of indices corresponding to the locations of ORFs in the <code>sequence</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/findgenes.jl#L3-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.orfgenerator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.orfgenerator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.orfgenerator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">orfgenerator(sequence::LongDNA; alternative_start::Bool=false, min_len = 6)</code></pre><p>Generate ORFs from the given DNA <code>sequence</code>.</p><p>This function generates ORFs from the forward and reverse complement strands of the <code>sequence</code> using the <code>locationgenerator</code> function.      It generates an ORF object for each range of indices returned by <code>locationgenerator</code>, and includes a <code>&#39;+&#39;</code> or <code>&#39;-&#39;</code> strand label          to indicate the strand from which the ORF was generated. The <code>sequence</code> argument must be a <code>LongDNA</code> object, which is a type          of DNA sequence with a longer maximum length than the <code>DNA</code> type.</p><p>Returns:     A generator expression that yields <code>ORF</code> objects corresponding to the ORFs in the <code>sequence</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/findgenes.jl#L31-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.simplecds_generator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.simplecds_generator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.simplecds_generator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">simplecds_generator(sequence::LongDNA; alternative_start::Bool=false, min_len = 6)
simplecds_generator(sequence::String; alternative_start::Bool=false, min_len = 6)</code></pre><p>A function to generete CDSs sequence out of a DNA sequence.</p><p>The <code>simplecds_generator</code> is a generator function that takes a <code>LongDNA</code> sequence and returns an iterator over the given sequence,     containing the coding sequences (CDSs) found in the sequence and the ORF.      It uses the <code>simplefind</code> function to find open reading frames (ORFs) in the sequence,      and then it extracts the actual CDS sequence from each ORF.      The function also searches the reverse complement of the sequence, so it finds CDSs on both strands.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/algorithms/simplefinder.jl#L53-L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.simplefind-Tuple{BioSequences.LongDNA}" href="#GeneFinder.simplefind-Tuple{BioSequences.LongDNA}"><code>GeneFinder.simplefind</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">simplefind(sequence::LongDNA; alternative_start::Bool=false)</code></pre><p>The simplest algorithm that finds ORFs in a DNA sequence.</p><p>The simplefind function takes a LongDNA sequence and returns a Vector{ORF} containing the ORFs found in the sequence. It searches the sequence for start codons (ATG) and stops when it finds a stop codon (TAG, TAA, or TGA), adding each ORF it finds to the vector. The function also searches the reverse complement of the sequence, so it finds ORFs on both strands.     This function has not ORFs size and overlapping condition contraints. Thus it might consider <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.         Extending the starting codons search to include ATG, GTG, and TTG.     In E. coli (K-12 strain), ATG is used in about 83 % of its genes, GTG in 14 % and TTG in 3 % of the cases @ref[axelson-fisk<em>comparative</em>2015]</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/algorithms/simplefinder.jl#L1-L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.simpleprot_generator-Tuple{BioSequences.LongDNA}" href="#GeneFinder.simpleprot_generator-Tuple{BioSequences.LongDNA}"><code>GeneFinder.simpleprot_generator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">simpleprot_generator(sequence::LongDNA; alternative_start::Bool=false, code::GeneticCode=BioSequences.standard_genetic_code, min_len = 6)
simpleprot_generator(sequence::String; alternative_start::Bool=false, code::GeneticCode=BioSequences.standard_genetic_code, min_len = 6)</code></pre><p>As its name suggest this generator function that iterates over the sequence to find proteins directly from a DNA sequence.      The <code>simpleprot_generator</code> function takes a <code>LongDNA</code> sequence and returns a <code>Vector{CDS}</code> containing the      coding sequences (CDSs) found in the sequence. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/algorithms/simplefinder.jl#L93-L100">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_bed-Tuple{String, BioSequences.LongDNA}" href="#GeneFinder.write_bed-Tuple{String, BioSequences.LongDNA}"><code>GeneFinder.write_bed</code></a> — <span class="docstring-category">Method</span></header><section><div><p>write<em>bed(file::String, seq::LongDNA; alternative</em>start::Bool=false)</p><p>Write BED data to a file.</p><p>Parameters:</p><ul><li>file: string; the file name to which the BED data will be written</li><li>seq: an instance of <code>LongDNA</code>, representing a long DNA sequence</li><li>alternative_start: boolean (optional); default is <code>false</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/io.jl#L1-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_cds-Tuple{String, BioSequences.LongDNA}" href="#GeneFinder.write_cds-Tuple{String, BioSequences.LongDNA}"><code>GeneFinder.write_cds</code></a> — <span class="docstring-category">Method</span></header><section><div><p>write<em>cds(file::String, seq::LongDNA; alternative</em>start=false, min_len = 6)</p><p>Write a file containing the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p>Parameters:</p><ul><li><code>file</code>: A string representing the file path and name where the CDSs should be written.</li><li><code>seq</code>: A <code>LongDNA</code> object representing the DNA sequence from which the CDSs should be extracted.</li></ul><p>Keyword Arguments:</p><ul><li><code>alternative_start</code>: A boolean value indicating whether alternative start codons should be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len</code>: An integer representing the minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/io.jl#L20-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_proteins-Tuple{String, BioSequences.LongDNA}" href="#GeneFinder.write_proteins-Tuple{String, BioSequences.LongDNA}"><code>GeneFinder.write_proteins</code></a> — <span class="docstring-category">Method</span></header><section><div><p>write<em>proteins(file::String, seq::LongDNA; alternative</em>start = false, code::GeneticCode = BioSequences.standard<em>genetic</em>code, min_len = 6)</p><p>Write a file containing the protein sequences encoded by the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p>Parameters:</p><ul><li><code>file</code>: A string representing the file path and name where the protein sequences should be written.</li><li><code>seq</code>: A <code>LongDNA</code> object representing the DNA sequence from which the CDSs and protein sequences should be extracted.</li></ul><p>Keyword Arguments:</p><ul><li><code>alternative_start</code>: A boolean value indicating whether alternative start codons should be used when identifying CDSs and translating them into protein sequences. Default is <code>false</code>.</li><li><code>code</code>: A <code>GeneticCode</code> object representing the genetic code that should be used to translate the CDSs into protein sequences. Default is the standard genetic code.</li><li><code>min_len</code>: An integer representing the minimum length that a protein sequence must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/496ccf99121fc94a3a8d6c3d4c876b43814c9963/src/io.jl#L41-L54">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../simplefinder/">« A first algorithm</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 10 January 2023 18:13">Tuesday 10 January 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
