<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneFinder.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/api/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../simplefinder/">Finding ORFs</a></li><li><a class="tocitem" href="../markovchains/">Towards Markov Chains</a></li><li><a class="tocitem" href="../markovchains-variant/">A variant test</a></li><li class="is-active"><a class="tocitem" href>API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/main/docs/src/api.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.CDS" href="#GeneFinder.CDS"><code>GeneFinder.CDS</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct CDS
    orf::ORF
    sequence::LongSubSeq{DNAAlphabet{4}}
end</code></pre><p>The <code>CDS</code> struct represents a coding sequence in a DNA sequence. It has three fields:</p><ul><li><code>orf</code>: is the basic composible type (<code>location::UnitRange{Int}</code>, strand::Char) displaying the location of the ORF and the associate strand: forward (&#39;+&#39;) or reverse (&#39;-&#39;)</li><li><code>sequence</code>: a <code>LongSequence{DNAAlphabet{4}}</code> sequence representing the actual sequence of the CDS</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L48-L58">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.GeneFeatures" href="#GeneFinder.GeneFeatures"><code>GeneFinder.GeneFeatures</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct GeneFeatures
    seqname::String
    start::Int64
    stop::Int64
    score::Float64
    strand::Char
    frame::&#39;.&#39;
    attribute::
end</code></pre><p>This is the main Gene struct, based on the fields that could be found in a GFF3, still needs to be defined correctly,     The idea is correct the frame and attributes that will have something like a possible list (id=Char;name=;locus_tag).     The <code>write</code> and <code>get</code> functions should have a dedicated method for this struct.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L7-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.ORF" href="#GeneFinder.ORF"><code>GeneFinder.ORF</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct ORF
    location::UnitRange{Int64}
    strand::Char
end</code></pre><p>The ORF struct represents an open reading frame in a DNA sequence. It has two fields: </p><ul><li><code>location</code>: which is a UnitRange{Int64} indicating the start and end locations of the ORF in the sequence</li><li><code>strand</code>:  is a Char type indicating whether the ORF is on the forward (&#39;+&#39;) or reverse (&#39;-&#39;) strand of the sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L32-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.Protein" href="#GeneFinder.Protein"><code>GeneFinder.Protein</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Protein
    sequence::LongSequence
    orf::ORF
end</code></pre><p>Similarly to the <code>CDS</code> struct, the <code>Protein</code> struct represents a encoded protein sequence in a DNA sequence.      It has three fields:</p><ul><li><code>orf</code>: is the basic composible type (<code>location::UnitRange{Int}</code>, strand::Char) of the sequence</li><li><code>sequence</code>: a <code>LongSequence</code> sequence representing the actual translated sequence of the CDS</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L64-L75">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.TCM" href="#GeneFinder.TCM"><code>GeneFinder.TCM</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">TCM(alphabet::Vector{DNA})</code></pre><p>A data structure for storing a DNA Transition Count Matrix (TCM). The TCM is a square matrix where each row and column corresponds to a nucleotide in the given <code>alphabet</code>. The value at position (i, j) in the matrix represents the number of times that nucleotide i is immediately followed by nucleotide j in a DNA sequence. </p><p>Fields:</p><ul><li><code>order::Dict{DNA, Int64}</code>: A dictionary that maps each nucleotide in the <code>alphabet</code> to its corresponding index in the matrix.</li><li><code>counts::Matrix{Int64}</code>: The actual matrix of counts.</li></ul><p>Internal function:</p><ul><li><code>TCM(alphabet::Vector{DNA})</code>: Constructs a new <code>TCM</code> object with the given <code>alphabet</code>. This function initializes the <code>order</code> field by creating a dictionary that maps each nucleotide in the <code>alphabet</code> to its corresponding index in the matrix. It also initializes the <code>counts</code> field to a matrix of zeros with dimensions <code>len x len</code>, where <code>len</code> is the length of the <code>alphabet</code>.</li></ul><p>Example usage:</p><pre><code class="language-julia hljs">alphabet = [DNA_A, DNA_C, DNA_G, DNA_T]
dtcm = TCM(alphabet)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L82-L99">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.TPM" href="#GeneFinder.TPM"><code>GeneFinder.TPM</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">TPM(alphabet::Vector{DNA})</code></pre><p>A data structure for storing a DNA Transition Probability Matrix (TPM). The TPM is a square matrix where each row and column corresponds to a nucleotide in the given <code>alphabet</code>. The value at position (i, j) in the matrix represents the probability of transitioning from nucleotide i to nucleotide j in a DNA sequence. </p><p>Fields:</p><ul><li><code>order::Dict{DNA, Int64}</code>: A dictionary that maps each nucleotide in the <code>alphabet</code> to its corresponding index in the matrix.</li><li><code>probabilities::Matrix{Float64}</code>: The actual matrix of probabilities.</li></ul><p>Example usage:</p><pre><code class="language-julia hljs">alphabet = [DNA_A, DNA_C, DNA_G, DNA_T]
dtpm = TPM(alphabet)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L118-L132">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.TransitionModel" href="#GeneFinder.TransitionModel"><code>GeneFinder.TransitionModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct TransitionModel</code></pre><p>The <code>TransitionModel</code> struct represents a transition model used in a sequence analysis. It consists of a transition probability matrix (tpm) and initial distribution probabilities.</p><p><strong>Fields</strong></p><ul><li><code>tpm::Matrix{Float64}</code>: The transition probability matrix, a matrix of type Float64 representing the probabilities of transitioning from one state to another.</li><li><code>initials::Matrix{Float64}</code>: The initial distribution probabilities, a matrix of type Float64 representing the probabilities of starting in each state.</li><li><code>n</code>: is the order of the transition model, or in other words the order of the resulted Markov chain.</li></ul><p><strong>Constructors</strong></p><ul><li><code>TransitionModel(tpm::Matrix{Float64}, initials::Matrix{Float64})</code>: Constructs a <code>TransitionModel</code> object with the provided transition probability matrix <code>tpm</code> and initial distribution probabilities <code>initials</code>.</li><li><code>TransitionModel(sequence::LongSequence{DNAAlphabet{4}})</code>: Constructs a <code>TransitionModel</code> object based on a given DNA sequence. The transition probability matrix is calculated using <code>transition_probability_matrix(sequence).probabilities</code>, and the initial distribution probabilities are calculated using <code>initial_distribution(sequence)</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/types.jl#L138-L154">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.cdsgenerator-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#GeneFinder.cdsgenerator-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>GeneFinder.cdsgenerator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">cdsgenerator(sequence::LongSequence{DNAAlphabet{4}}; kwargs...)
cdsgenerator(sequence::String; kwargs...)</code></pre><p>A function to generete CDSs sequence out of a DNA sequence.</p><p>The <code>cdsgenerator</code> is a generator function that takes a <code>LongSequence{DNAAlphabet{4}}</code> sequence and returns an iterator over the given sequence,     containing the coding sequences (CDSs) found in the sequence and the ORF.      It uses the <code>findorfs</code> function to find open reading frames (ORFs) in the sequence,      and then it extracts the actual CDS sequence from each ORF, returining both.      The function also searches the reverse complement of the sequence, so it finds CDSs on both strands.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/findgenes.jl#L3-L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.dinucleotides-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}" href="#GeneFinder.dinucleotides-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}"><code>GeneFinder.dinucleotides</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">dinucleotides(sequence::LongNucOrView{4})</code></pre><p>Compute the transition counts of each dinucleotide in a given DNA sequence.</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongSequence{DNAAlphabet{4}}</code>: a <code>LongSequence{DNAAlphabet{4}}</code> object representing the DNA sequence.</li></ul><p><strong>Keywords</strong></p><ul><li><code>extended_alphabet::Bool=false</code>: If true will pass the extended alphabet of DNA to search</li></ul><p><strong>Returns</strong></p><p>A dictionary with keys being <code>LongSequence{DNAAlphabet{4}}</code> objects representing the dinucleotides, and values being the number of occurrences of each dinucleotide in the sequence.</p><p><strong>Example</strong></p><pre><code class="nohighlight hljs">seq = dna&quot;AGCTAGCTAGCT&quot;

dinucleotides(seq)

Dict{LongSequence{DNAAlphabet{4}}, Int64} with 16 entries:
  GG =&gt; 0
  TC =&gt; 0
  GC =&gt; 3
  CG =&gt; 0
  CC =&gt; 0
  AG =&gt; 3
  TT =&gt; 0
  AC =&gt; 0
  TA =&gt; 2
  GT =&gt; 0
  GA =&gt; 0
  CT =&gt; 3
  CA =&gt; 0
  AT =&gt; 0
  AA =&gt; 0
  TG =&gt; 0</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/helpers.jl#L69-L110">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.fasta_to_dna-Tuple{String}" href="#GeneFinder.fasta_to_dna-Tuple{String}"><code>GeneFinder.fasta_to_dna</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fasta_to_dna(input::String)</code></pre><p>Converts a FASTA formatted file (even if it is a multi-fasta) to an array of <code>LongSequence{DNAAlphabet{4}}</code> objects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/helpers.jl#L2-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.findorfs-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#GeneFinder.findorfs-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>GeneFinder.findorfs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findorfs(sequence::LongSequence{DNAAlphabet{4}}; kwargs...)::Vector{ORF}
findorfs(sequence::String; kwargs...)::Vector{ORF}</code></pre><p>A simple implementation that finds ORFs in a DNA sequence.</p><p>The <code>findorfs</code> function takes a LongSequence{DNAAlphabet{4}} sequence and returns a Vector{ORF} containing the ORFs found in the sequence.      It searches entire regularly expressed CDS, adding each ORF it finds to the vector. The function also searches the reverse complement of the sequence, so it finds ORFs on both strands.         Extending the starting codons with the <code>alternative_start = true</code> will search for ATG, GTG, and TTG.     Some studies have shown that in <em>E. coli</em> (K-12 strain), ATG, GTG and TTG are used 83 %, 14 % and 3 % respectively.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This function has not ORFs scoring scheme. Thus it might consider <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</p></div></div><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/algorithms/findorfs.jl#L14-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.getcds-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#GeneFinder.getcds-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>GeneFinder.getcds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getcds(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...)
getcds(input::String, output::String; kwargs...) ## for strings per se</code></pre><p>This function will take a <code>LongSequence{DNAAlphabet{4}}</code> or <code>String</code> sequence and by means of the <code>findorfs()</code> function will push <code>LongSubSeq{DNAAlphabet{4}}</code> into a <code>Vector{}</code></p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/findgenes.jl#L58-L68">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.getproteins-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#GeneFinder.getproteins-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>GeneFinder.getproteins</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getproteins(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...)
getproteins(input::String, output::String; kwargs...)</code></pre><p>Similar to <code>getcds()</code> function, it will take a <code>LongSequence{DNAAlphabet{4}}</code> or <code>String</code> sequence and by means of the <code>findorfs()</code> and the <code>translate()</code> function will push <code>LongAA</code>s into a <code>Vector</code></p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/findgenes.jl#L163-L173">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.hasprematurestop-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}" href="#GeneFinder.hasprematurestop-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}"><code>GeneFinder.hasprematurestop</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">hasprematurestop(sequence::LongNucOrView{4})::Bool</code></pre><p>Determine whether the <code>sequence</code> of type <code>LongSequence{DNAAlphabet{4}}</code> contains a premature stop codon.</p><p>Returns a boolean indicating whether the <code>sequence</code> has more than one stop codon.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/helpers.jl#L44-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.iscoding" href="#GeneFinder.iscoding"><code>GeneFinder.iscoding</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">iscoding(
    sequence::LongSequence{DNAAlphabet{4}}, 
    codingmodel::TransitionModel, 
    noncodingmodel::TransitionModel,
    η::Float64 = 1e-5
    )</code></pre><p>Check if a given DNA sequence is likely to be coding based on a log-odds ratio.     The log-odds ratio is a statistical measure used to assess the likelihood of a sequence being coding or non-coding. It compares the probability of the sequence generated by a coding model to the probability of the sequence generated by a non-coding model. If the log-odds ratio exceeds a given threshold (<code>η</code>), the sequence is considered likely to be coding.     It is formally described as a decision rule:</p><p class="math-container">\[S(X) = \log \left( \frac{{P_C(X_1=i_1, \ldots, X_T=i_T)}}{{P_N(X_1=i_1, \ldots, X_T=i_T)}} \right) \begin{cases} &gt; \eta &amp; \Rightarrow \text{{coding}} \\ &lt; \eta &amp; \Rightarrow \text{{noncoding}} \end{cases}\]</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongSequence{DNAAlphabet{4}}</code>: The DNA sequence to be evaluated.</li><li><code>codingmodel::TransitionModel</code>: The transition model for coding regions.</li><li><code>noncodingmodel::TransitionModel</code>: The transition model for non-coding regions.</li><li><code>η::Float64 = 1e-5</code>: The threshold value for the log-odds ratio (default: 1e-5).</li></ul><p><strong>Returns</strong></p><ul><li><code>true</code> if the sequence is likely to be coding.</li><li><code>false</code> if the sequence is likely to be non-coding.</li></ul><p><strong>Raises</strong></p><ul><li><code>ErrorException</code>: if the length of the sequence is not divisible by 3.</li><li><code>ErrorException</code>: if the sequence contains a premature stop codon.</li></ul><p><strong>Example</strong></p><pre><code class="nohighlight hljs">sequence = LondDNA(&quot;ATGGCATCTAG&quot;)
codingmodel = TransitionModel()
noncodingmodel = TransitionModel()
iscoding(sequence, codingmodel, noncodingmodel)  # Returns: true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/biomarkovchains.jl#L265-L303">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.locationiterator-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#GeneFinder.locationiterator-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>GeneFinder.locationiterator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">locationiterator(sequence::LongSequence{DNAAlphabet{4}}; alternative_start::Bool=false)</code></pre><p>This is an iterator function that uses regular expressions to search the entire CDS (instead of start and stop codons) in a <code>LongSequence{DNAAlphabet{4}}</code> sequence.     It uses an anonymous function that will find the first regularly expressed CDS. Then using this anonymous function it creates an iterator that will apply it until there is no other CDS.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/algorithms/findorfs.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.nucleotidefreqs-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}" href="#GeneFinder.nucleotidefreqs-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}"><code>GeneFinder.nucleotidefreqs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">nucleotidefreqs(sequence::LongSequence{DNAAlphabet{4}}) -&gt; Dict{DNA, Float64}</code></pre><p>Calculate the frequency of each nucleotide in a DNA sequence.</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongSequence{DNAAlphabet{4}}</code>: A <code>LongSequence{DNAAlphabet{4}}</code> sequence.</li></ul><p><strong>Returns</strong></p><p>A dictionary with each nucleotide in the sequence as a key, and its frequency as a value.</p><p><strong>Example</strong></p><pre><code class="nohighlight hljs">seq = dna&quot;CCTCCCGGACCCTGGGCTCGGGAC&quot;

nucleotidefreqs(seq)

Dict{DNA, Float64} with 4 entries:
DNA_T =&gt; 0.125
DNA_A =&gt; 0.0833333
DNA_G =&gt; 0.333333
DNA_C =&gt; 0.458333</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/helpers.jl#L13-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.proteingenerator-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#GeneFinder.proteingenerator-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>GeneFinder.proteingenerator</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">proteingenerator(sequence::LongSequence{DNAAlphabet{4}}; kwargs...)
proteingenerator(sequence::String; kwargs...)</code></pre><p>As its name suggest this generator function iterates over the sequence to find proteins directly from a DNA sequence.      The <code>proteingenerator</code> function takes a <code>LongSequence{DNAAlphabet{4}}</code> sequence and returns a <code>Vector{CDS}</code> containing the      coding sequences (CDSs) found in the sequence and the associated ORF.</p><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/findgenes.jl#L101-L114">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.sequenceprobability-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}, TransitionModel}" href="#GeneFinder.sequenceprobability-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}, TransitionModel}"><code>GeneFinder.sequenceprobability</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequenceprobability(sequence::LongNucOrView{4}, tpm::Matrix{Float64}, initials=Vector{Float64})</code></pre><p>Compute the probability of a given sequence using a transition probability matrix and the initial probabilities distributions.</p><p class="math-container">\[P(X_1 = i_1, \ldots, X_T = i_T) = \pi_{i_1}^{T-1} \prod_{t=1}^{T-1} a_{i_t, i_{t+1}}\]</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongNucOrView{4}</code>: The input sequence of nucleotides.</li><li><code>tm::TransitionModel</code> is the actual data structure composed of a <code>tpm::Matrix{Float64}</code> the transition probability matrix and <code>initials=Vector{Float64}</code> the initial state probabilities.</li></ul><p><strong>Returns</strong></p><ul><li><code>probability::Float64</code>: The probability of the input sequence.</li></ul><p><strong>Example</strong></p><pre><code class="nohighlight hljs">mainseq = LongDNA{4}(&quot;CCTCCCGGACCCTGGGCTCGGGAC&quot;)

tpm = transition_probability_matrix(mainseq)
    
    4×4 Matrix{Float64}:
    0.0   1.0    0.0    0.0
    0.0   0.5    0.2    0.3
    0.25  0.125  0.625  0.0
    0.0   0.667  0.333  0.0

initials = initial_distribution(mainseq)

    1×4 Matrix{Float64}:
    0.0869565  0.434783  0.347826  0.130435
    
tm = transition_model(tpm, initials)
    - Transition Probability Matrix (Size: 4 × 4):
    0.0	1.0	0.0	0.0
    0.0	0.5	0.2	0.3
    0.25	0.125	0.625	0.0
    0.0	0.667	0.333	0.0
    - Initials (Size: 1 × 4):
    0.087	0.435	0.348	0.13
    - order: 1

newseq = LondDNA(&quot;CCTG&quot;)

    4nt DNA Sequence:
    CCTG


sequenceprobability(newseq, tm)
    
    0.0217</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/biomarkovchains.jl#L169-L223">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.transition_count_matrix-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}" href="#GeneFinder.transition_count_matrix-Tuple{Union{BioSequences.LongSubSeq{&lt;:BioSequences.NucleicAcidAlphabet{4}}, BioSequences.LongNuc{4, &lt;:BioSequences.NucleicAcidAlphabet{4}}}}"><code>GeneFinder.transition_count_matrix</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">transition_count_matrix(sequence::LongSequence{DNAAlphabet{4}})</code></pre><p>Compute the transition count matrix (TCM) of a given DNA sequence.</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongSequence{DNAAlphabet{4}}</code>: a <code>LongSequence{DNAAlphabet{4}}</code> object representing the DNA sequence.</li></ul><p><strong>Keywords</strong></p><ul><li><code>extended_alphabet::Bool=false</code>: If true will pass the extended alphabet of DNA to search</li></ul><p><strong>Returns</strong></p><p>A <code>TCM</code> object representing the transition count matrix of the sequence.</p><p><strong>Example</strong></p><pre><code class="nohighlight hljs">seq = LongDNA{4}(&quot;AGCTAGCTAGCT&quot;)

tcm = transition_count_matrix(seq)

GeneFinder.TCM{Dict{DNA, Int64}, Matrix{Int64}:
   A C G T
A  0 0 3 0
C  0 0 0 3
G  0 3 0 0
T  2 0 0 0
</code></pre><pre><code class="nohighlight hljs"></code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/biomarkovchains.jl#L1-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.transition_model" href="#GeneFinder.transition_model"><code>GeneFinder.transition_model</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">transition_model(sequence::LongNucOrView{4}, n::Int64=1)</code></pre><p>Constructs a transition model based on the given DNA sequence and transition order.</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongNucOrView{4}</code>: A DNA sequence represented as a <code>LongNucOrView{4}</code> object.</li><li><code>n::Int64 (optional)</code>: The transition order (default: 1).</li></ul><p><strong>Returns</strong></p><p>A <code>TransitionModel</code> object representing the transition model.</p><hr/><pre><code class="nohighlight hljs">transition_model(tpm::Matrix{Float64}, initials::Matrix{Float64}, n::Int64=1)</code></pre><p>Builds a transtition model based on the transition probability matrix and the initial distributions. It can also calculates higer orders of the model if <code>n</code> is changed.</p><p><strong>Arguments</strong></p><ul><li><code>tpm::Matrix{Float64}</code>: the transition probability matrix <code>TPM</code></li><li><code>initials::Matrix{Float64}</code>: the initial distributions of the model.</li><li><code>n::Int64 (optional)</code>: The transition order (default: 1).</li></ul><p><strong>Returns</strong></p><p>A <code>TransitionProbabilityMatrix</code> object representing the transition probability matrix.</p><p><strong>Example</strong></p><pre><code class="nohighlight hljs">sequence = LongDNA{4}(&quot;ACTACATCTA&quot;)

model = transition_model(sequence, 2)
TransitionModel:
  - Transition Probability Matrix (Size: 4 × 4):
    0.444	0.111	0.0	0.444
    0.444	0.444	0.0	0.111
    0.0	0.0	0.0	0.0
    0.111	0.444	0.0	0.444
  - Initials (Size: 1 × 4):
    0.333	0.333	0.0	0.333
  - order: 2</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/biomarkovchains.jl#L117-L158">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.transition_probability_matrix" href="#GeneFinder.transition_probability_matrix"><code>GeneFinder.transition_probability_matrix</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">transition_probability_matrix(sequence::LongSequence{DNAAlphabet{4}})</code></pre><p>Compute the transition probability matrix (TPM) of a given DNA sequence. Formally it construct <span>$\hat{A}$</span> where: </p><p class="math-container">\[a_{ij} = P(X_t = j \mid X_{t-1} = i) = \frac{{P(X_{t-1} = i, X_t = j)}}{{P(X_{t-1} = i)}}\]</p><p><strong>Arguments</strong></p><ul><li><code>sequence::LongNucOrView{4}</code>: a <code>LongNucOrView{4}</code> object representing the DNA sequence.</li><li><code>n::Int64=1</code>: The order of the Markov model. That is the <span>$\hat{A}^{n}$</span></li></ul><p><strong>Keywords</strong></p><ul><li><code>extended_alphabet::Bool=false</code>: If true will pass the extended alphabet of DNA to search</li></ul><p><strong>Returns</strong></p><p>A <code>TPM</code> object representing the transition probability matrix of the sequence.</p><p><strong>Example</strong></p><pre><code class="nohighlight hljs">seq = dna&quot;AGCTAGCTAGCT&quot;

tpm = transition_probability_matrix(seq)

GeneFinder.tpm{Dict{DNA, Int64}, Matrix{Float64}:
   A   C   G   T
A  0.0 0.0 1.0 0.0
C  0.0 0.0 0.0 1.0
G  0.0 1.0 0.0 0.0
T  1.0 0.0 0.0 0.0</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/biomarkovchains.jl#L53-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_bed-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, String}" href="#GeneFinder.write_bed-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, String}"><code>GeneFinder.write_bed</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_bed(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...)
write_bed(input::String, output::String; kwargs...)</code></pre><p>Write BED data to a file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/io.jl#L1-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_cds-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, String}" href="#GeneFinder.write_cds-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, String}"><code>GeneFinder.write_cds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_cds(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...)
write_cds(input::String, output::String; kwargs...)</code></pre><p>Write a file containing the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start</code>: A boolean value indicating whether alternative start codons should be used when identifying CDSs. Default is <code>false</code>.</li><li><code>min_len</code>: An integer representing the minimum length that a CDS must have in order to be included in the output file. Default is <code>6</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/io.jl#L33-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_gff-Tuple{String, String}" href="#GeneFinder.write_gff-Tuple{String, String}"><code>GeneFinder.write_gff</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_gff(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...)
write_gff(input::String, output::String; kwargs...)</code></pre><p>Write GFF data to a file.</p><p><strong>Keywords</strong></p><ul><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/io.jl#L108-L118">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GeneFinder.write_proteins-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, String}" href="#GeneFinder.write_proteins-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, String}"><code>GeneFinder.write_proteins</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_proteins(input::LongSequence{DNAAlphabet{4}}, output::String; kwargs...)</code></pre><p>Write the protein sequences encoded by the coding sequences (CDSs) of a given DNA sequence to the specified file.</p><p><strong>Keywords</strong></p><ul><li><code>code::GeneticCode=BioSequences.standard_genetic_code</code>: The genetic code by which codons will be translated. See <code>BioSequences.ncbi_trans_table</code> for more info. </li><li><code>alternative_start::Bool=false</code>: If true will pass the extended start codons to search. This will increase 3x the exec. time.</li><li><code>min_len::Int64=6</code>:  Length of the allowed ORF. Default value allow <code>aa&quot;M*&quot;</code> a posible encoding protein from the resulting ORFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/714e4ae9d4ae243db6935bf0492bc2736120c9a3/src/io.jl#L65-L75">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../markovchains-variant/">« A variant test</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 4 July 2023 15:43">Tuesday 4 July 2023</span>. Using Julia version 1.9.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
