<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Towards Markov Chains Â· GeneFinder.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://camilogarciabotero.github.io/GeneFinder.jl/markovchains/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GeneFinder.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../simplefinder/">Finding ORFs</a></li><li class="is-active"><a class="tocitem" href>Towards Markov Chains</a><ul class="internal"><li><a class="tocitem" href="#Markov-chain-BioSequences"><span>Markov chain BioSequences</span></a></li><li><a class="tocitem" href="#References"><span>References</span></a></li></ul></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Towards Markov Chains</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Towards Markov Chains</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/camilogarciabotero/GeneFinder.jl/blob/main/docs/src/markovchains.md#" title="Edit on GitHub"><span class="docs-icon fab">ï‚›</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="DNA-as-a-Markov-chain"><a class="docs-heading-anchor" href="#DNA-as-a-Markov-chain">DNA as a Markov chain</a><a id="DNA-as-a-Markov-chain-1"></a><a class="docs-heading-anchor-permalink" href="#DNA-as-a-Markov-chain" title="Permalink"></a></h1><p>Several packages (e.g.Â <a href=".">MarkovChainsHammer</a>, <a href=".">Markovians</a>, etc.) in the Julia ecosystem have been implemented to work with Markov chains with a <em>state space</em> of integers, those could be efficient in many ways, but they are clumsy to work with a specialized biological types as in the <code>BioJulia</code> ecosystem. Therefore, in the <code>GeneFinder</code> package we dedicated some implementations to work with <code>BioSequence</code> types so that we can expand the functionality in an efficient way (see complete <a href=".">API</a>).</p><p>One important step towards many gene finding algorithms is to represent a DNA sequence as a Markov chain. In this representation a DNA sequence of a reduced alphabet ğ’œâ€„=â€„{<em>A</em>,â€†<em>C</em>,â€†<em>G</em>,â€†<em>T</em>} is draw as a four-vertex graph, where each letter of ğ’œ is a <em>state (vertex)</em> and the edges of the graph represent <em>transitions</em> from one nucleotide to another in a sequence (e.g.Â <em>A</em>â€„â†’â€„<em>T</em> represent a single nucleotide to nucleotide transition). This is also considered more specifically as a Discrete Markov chain (Axelson-Fisk 2015). The complete set of transitions and states of a DNA sequence of alphabet ğ’œ could be seen as &lt;a href=&quot;#fig-dna-markov&quot; class=&quot;quarto-xref&quot;&gt;FigureÂ 1&lt;/a&gt;.</p><p>&lt;img src=&quot;../assets/nucleotide-markov-chain.svg&quot; id=&quot;fig-dna-markov&quot; data-fig-align=&quot;center&quot; alt=&quot;FigureÂ 1: DNA sequence as a Markov chain with a DNA alphabet \mathscr{A} = {A, C, G, T}&quot; /&gt;</p><p>More formally a Markov chain is a random process where each state is a random variable <em>X</em>&lt;sub&gt;<em>t</em>&lt;/sub&gt; where <em>t</em>â€„âˆˆâ€„<em>T</em> is a discrete time in a finite sequence <em>T</em> and the probability to jump from one state into another is <em>only dependent of the current state.</em> Therefore a definition of this <em>Markov property</em> is given by:</p><p>&lt;span id=&quot;eq-markov-property&quot;&gt; <em>P</em>(<em>X</em>&lt;sub&gt;<em>t</em>&lt;/sub&gt;=<em>j</em>|<em>X</em>&lt;sub&gt;<em>t</em>â€…âˆ’â€…1&lt;/sub&gt;=<em>i</em>)â€Šâ€â€(1) &lt;/span&gt;</p><p>where <em>i</em>,â€†<em>j</em>â€„âˆˆâ€„ğ’œ . This property led us to generalize a way to calculate the probability of a sequence <em>T</em> from a process <em>X</em>&lt;sub&gt;1&lt;/sub&gt;...<em>X</em>&lt;sub&gt;<em>T</em>&lt;/sub&gt; where each random variable is a nucleotide from ğ’œ so that:</p><p>&lt;span id=&quot;eq-seq-probability&quot;&gt; $ P(X_{1} = i_{1},...,X_{T} = i_{T}) = P(X_{1} = i_{1}) \prod_{t=2}^{T} P(X_{t} = i_{t} | X_{tâˆ’1}) = i_{tâˆ’1}).  \qquad(2)$ &lt;/span&gt;</p><p>Note that &lt;a href=&quot;#eq-seq-probability&quot; class=&quot;quarto-xref&quot;&gt;EquationÂ 2&lt;/a&gt; has two terms, a initial probability <em>P</em>(<em>X</em>&lt;sub&gt;1&lt;/sub&gt;=<em>i</em>&lt;sub&gt;1&lt;/sub&gt;) and the the product of all transitions beginning at <em>t</em>â€„=â€„2.</p><h2 id="Markov-chain-BioSequences"><a class="docs-heading-anchor" href="#Markov-chain-BioSequences">Markov chain BioSequences</a><a id="Markov-chain-BioSequences-1"></a><a class="docs-heading-anchor-permalink" href="#Markov-chain-BioSequences" title="Permalink"></a></h2><p>We can now calculate a transition matrix from a <code>LongDNA</code> sequence using <code>transition_probability_matrix</code> method</p><pre><code class="language-julia hljs">using BioSequences, GeneFinder

genome = randdnaseq(10^6)

cds = getcds(genome, min_len = 64)[1]

cds</code></pre><pre><code class="nohighlight hljs">129nt DNA Sequence:
ATGTATTACAAAGTTACTTATGCTTCCTCGTTCTACGTTâ€¦TCTGGGAGTGAACGCGTGAGCCCATATTTTATGCTTTAA</code></pre><p>The <code>cds</code> object is a an ORF with the potential to encode a CDS from the randomly generated <code>genome</code>. To see what is the transition probabilities and the initial distribution, we can use build the Transition Model using simply the constructor <code>TransitionModel</code></p><pre><code class="language-julia hljs">TransitionModel(cds)</code></pre><pre><code class="nohighlight hljs">TransitionModel:
  - Transition Probability Matrix (Size: 4 Ã— 4):
    0.235   0.235   0.235   0.294   
    0.2 0.24    0.28    0.28    
    0.333   0.185   0.148   0.333   
    0.286   0.143   0.19    0.381   
  - Initials (Size: 1 Ã— 4):
    0.266   0.195   0.211   0.328   
  - order: 1</code></pre><h2 id="References"><a class="docs-heading-anchor" href="#References">References</a><a id="References-1"></a><a class="docs-heading-anchor-permalink" href="#References" title="Permalink"></a></h2><p>Axelson-Fisk, Marina. 2015. <em>Comparative Gene Finding</em>. Vol. 20. Computational Biology. London: Springer London. <a href="http://link.springer.com/10.1007/978-1-4471-6693-1">http://link.springer.com/10.1007/978-1-4471-6693-1</a>.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../simplefinder/">Â« Finding ORFs</a><a class="docs-footer-nextpage" href="../api/">API Â»</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Wednesday 28 June 2023 00:34">Wednesday 28 June 2023</span>. Using Julia version 1.9.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
